{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cypher Queries for Determining Regulatory Paths\n",
    "*Núria Queralt Rosinach, Andrew Su*\n",
    "\n",
    "**Freeze Lab meeting, January 2019**\n",
    "\n",
    "## Overview\n",
    "NGLY1 - AQP1 **regulatory review** (*NGLY1 v3.1*)\n",
    "\n",
    "## Servers \n",
    "\n",
    "    * Local: bolt://kylo.scripps.edu:7689\n",
    "    * AWS: bolt://52.87.232.110:7689"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from neo4j.v1 import GraphDatabase, basic_auth\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)  # or 1000\n",
    "pd.set_option('display.max_rows', None)  # or 1000\n",
    "pd.set_option('display.max_colwidth', -1)  # or 199"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def runQuery( driver, query ):\n",
    "    '''\n",
    "    This function runs the query onto the database and returns the result.\n",
    "    in: cypher query string\n",
    "    out: neo4j query result object\n",
    "    '''\n",
    "    \n",
    "    with driver.session() as session:\n",
    "        result = session.run('' + query + '')\n",
    "        \n",
    "    return result\n",
    "\n",
    "\n",
    "def parseNode( node ):\n",
    "    '''\n",
    "    This function parses the information gathered in the node data structure object resulting after querying neo4j.\n",
    "        in: node record neo4j object\n",
    "        out: node as dict\n",
    "    '''\n",
    "    \n",
    "    n = dict()\n",
    "    n[\"idx\"] = int(node.id)\n",
    "    n[\"type\"] = list(node.labels)[0]\n",
    "    n[\"id\"] = str(node.properties['id'])\n",
    "    n[\"preflabel\"] = str(node.properties['preflabel'])\n",
    "    n[\"name\"] = str(node.properties['name'])\n",
    "    n[\"description\"] = str(node.properties['description'])\n",
    "\n",
    "    return n\n",
    "\n",
    "\n",
    "def parsePath( path ):\n",
    "    '''\n",
    "    This function parsers the information gathered in the path data structure object resulting after querying neo4j.\n",
    "        in: path record neo4j object\n",
    "        out: path as dict\n",
    "    '''\n",
    "    \n",
    "    out = {}\n",
    "    out['Nodes'] = []\n",
    "    for node in path['path'].nodes:\n",
    "        n = {}\n",
    "        n['idx'] = int(node.id)\n",
    "        n['type'] = list(node.labels)[0]\n",
    "        n['id'] = str(node.properties['id'])\n",
    "        n['preflabel'] = str(node.properties['preflabel'])\n",
    "        n['name'] = str(node.properties['name'])\n",
    "        n['description'] = str(node.properties['description'])\n",
    "        out['Nodes'].append(n)\n",
    "    out['Edges'] = []\n",
    "    for edge in path['path'].relationships:\n",
    "        e = {}\n",
    "        e['idx'] = int(edge.id)\n",
    "        e['start_node'] = int(edge.start)\n",
    "        e['end_node'] = int(edge.end)\n",
    "        e['type'] = str(edge.type)\n",
    "        e['property_label'] = str(edge.properties['property_label'])\n",
    "        e['property_uri'] = str(edge.properties['property_uri'])\n",
    "        e['reference_uri'] = str(edge.properties['reference_uri'])\n",
    "        e['reference_date'] = str(edge.properties['reference_date'])\n",
    "        e['reference_supporting_text'] = str(edge.properties['reference_supporting_text'])\n",
    "        out['Edges'].append(e)\n",
    "        \n",
    "    return out"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initialize neo4j"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "driver = GraphDatabase.driver(\"bolt://kylo.scripps.edu:7689\", auth=basic_auth(\"neo4j\", \"xena\"))\n",
    "#driver = GraphDatabase.driver(\"bolt://52.87.232.110:7689\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Question\n",
    "## Query topology graph\n",
    "## Table of summary\n",
    "## Graph of paths\n",
    "## Explore paths=> executable cypher query"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TRANSCRIPTOME ANALYSIS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Common TFs for human RNA transcriptome\n",
    "* **Query web 1: < NGLY1 -[Human RNA expression (Freeze)]- gene <-- TF > order by freq**\n",
    "    * ngly1-[interacts with]-gene<-tf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(857, 5)\n",
      "CPU times: user 30.9 ms, sys: 286 µs, total: 31.2 ms\n",
      "Wall time: 10.5 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)-[i1:`RO:0002434`]-(gene:GENE)<-[i2:`RO:0002434`]-(tf:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,i2,tf,gene,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 \n",
    "        \n",
    "        AND toLower(i1.reference_supporting_text) =~ '.*freeze.*' \n",
    "        \n",
    "        AND toLower(i2.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN DISTINCT tf.id as id, tf.preflabel as symbol, tf.name as name, tf.description as description, count(distinct gene.preflabel) as freq\n",
    "        \n",
    "        ORDER BY freq DESC\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "out_l = list()\n",
    "for record in result:\n",
    "    out_l.append({\n",
    "        'TF_id': record['id'], \n",
    "        'TF_symbol': record['symbol'],\n",
    "        'TF_name': record['name'],\n",
    "        'TF_description': record['description'],\n",
    "        'Frequency': record['freq']\n",
    "    })\n",
    "    \n",
    "res_df = pd.DataFrame(out_l)\n",
    "print(res_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Frequency</th>\n",
       "      <th>TF_id</th>\n",
       "      <th>TF_symbol</th>\n",
       "      <th>TF_name</th>\n",
       "      <th>TF_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>471</td>\n",
       "      <td>HGNC:11205</td>\n",
       "      <td>SP1</td>\n",
       "      <td>Sp1 transcription factor</td>\n",
       "      <td>The protein encoded by this gene is a zinc finger transcription factor that binds to GC-rich motifs of many promoters. The encoded protein is involved in many cellular processes, including cell differentiation, cell growth, apoptosis, immune responses, response to DNA damage, and chromatin remodeling. Post-translational modifications such as phosphorylation, acetylation, glycosylation, and proteolytic processing significantly affect the activity of this protein, which can be an activator or a repressor. Three transcript variants encoding different isoforms have been found for this gene. [provided by RefSeq, Nov 2014].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>403</td>\n",
       "      <td>HGNC:6551</td>\n",
       "      <td>LEF1</td>\n",
       "      <td>lymphoid enhancer binding factor 1</td>\n",
       "      <td>This gene encodes a transcription factor belonging to a family of proteins that share homology with the high mobility group protein-1. The protein encoded by this gene can bind to a functionally important site in the T-cell receptor-alpha enhancer, thereby conferring maximal enhancer activity. This transcription factor is involved in the Wnt signaling pathway, and it may function in hair cell differentiation and follicle morphogenesis. Mutations in this gene have been found in somatic sebaceous tumors. This gene has also been linked to other cancers, including androgen-independent prostate cancer. Alternative splicing results in multiple transcript variants. [provided by RefSeq, Oct 2009].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>305</td>\n",
       "      <td>HGNC:6914</td>\n",
       "      <td>MAZ</td>\n",
       "      <td>MYC associated zinc finger protein</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>287</td>\n",
       "      <td>HGNC:7139</td>\n",
       "      <td>FOXO4</td>\n",
       "      <td>forkhead box O4</td>\n",
       "      <td>This gene encodes a member of the O class of winged helix/forkhead transcription factor family. Proteins encoded by this class are regulated by factors involved in growth and differentiation indicating they play a role in these processes. A translocation involving this gene on chromosome X and the homolog of the Drosophila trithorax gene, encoding a DNA binding protein, located on chromosome 11 is associated with leukemia. Multiple transcript variants encoding different isoforms have been found for this gene. [provided by RefSeq, Jan 2010].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>283</td>\n",
       "      <td>NCBIGene:26196534</td>\n",
       "      <td>E12</td>\n",
       "      <td>transcription regulator</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Frequency              TF_id TF_symbol                             TF_name  \\\n",
       "0  471        HGNC:11205         SP1       Sp1 transcription factor             \n",
       "1  403        HGNC:6551          LEF1      lymphoid enhancer binding factor 1   \n",
       "2  305        HGNC:6914          MAZ       MYC associated zinc finger protein   \n",
       "3  287        HGNC:7139          FOXO4     forkhead box O4                      \n",
       "4  283        NCBIGene:26196534  E12       transcription regulator              \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               TF_description  \n",
       "0  The protein encoded by this gene is a zinc finger transcription factor that binds to GC-rich motifs of many promoters. The encoded protein is involved in many cellular processes, including cell differentiation, cell growth, apoptosis, immune responses, response to DNA damage, and chromatin remodeling. Post-translational modifications such as phosphorylation, acetylation, glycosylation, and proteolytic processing significantly affect the activity of this protein, which can be an activator or a repressor. Three transcript variants encoding different isoforms have been found for this gene. [provided by RefSeq, Nov 2014].                                                                           \n",
       "1  This gene encodes a transcription factor belonging to a family of proteins that share homology with the high mobility group protein-1. The protein encoded by this gene can bind to a functionally important site in the T-cell receptor-alpha enhancer, thereby conferring maximal enhancer activity. This transcription factor is involved in the Wnt signaling pathway, and it may function in hair cell differentiation and follicle morphogenesis. Mutations in this gene have been found in somatic sebaceous tumors. This gene has also been linked to other cancers, including androgen-independent prostate cancer. Alternative splicing results in multiple transcript variants. [provided by RefSeq, Oct 2009].  \n",
       "2  NA                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \n",
       "3  This gene encodes a member of the O class of winged helix/forkhead transcription factor family. Proteins encoded by this class are regulated by factors involved in growth and differentiation indicating they play a role in these processes. A translocation involving this gene on chromosome X and the homolog of the Drosophila trithorax gene, encoding a DNA binding protein, located on chromosome 11 is associated with leukemia. Multiple transcript variants encoding different isoforms have been found for this gene. [provided by RefSeq, Jan 2010].                                                                                                                                                          \n",
       "4  NA                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Summary table\n",
    "res_df[['Frequency', 'TF_id', 'TF_symbol', 'TF_name', 'TF_description']].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RNA gene clusters by TF and Pathway (pathways associated with gene sets)\n",
    "    * Cluster RNA genes-common TF clusters by pathway (pathways associated with RNA genes in gene sets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(850, 6)\n",
      "CPU times: user 166 ms, sys: 0 ns, total: 166 ms\n",
      "Wall time: 42 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 20\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)-[i1:`RO:0002434`]->(g:GENE)<-[i2:`RO:0002434`]-(tf:GENE), (g:GENE)-[i]-(pw:PHYS)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,i2,g,tf,i,pw,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 \n",
    "        \n",
    "        AND toLower(i1.reference_supporting_text) =~ '.*freeze.*' \n",
    "        \n",
    "        AND toLower(i2.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "        \n",
    "        AND toLower(i.property_label) <> 'enables'\n",
    "\n",
    "        RETURN DISTINCT tf.preflabel as TF_symbol, tf.name as TF_name, \n",
    "                        collect(DISTINCT g.preflabel) as geneset, count(distinct g.preflabel) as genes,\n",
    "                        collect(DISTINCT pw.preflabel) as pathway, count(distinct pw.preflabel) as pathways\n",
    "        \n",
    "        ORDER BY genes DESC\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "out_l = list()\n",
    "for record in result:\n",
    "    out_l.append({'TF symbol': record['TF_symbol'],\n",
    "                  'TF name': record['TF_name'],\n",
    "                  'Target geneset': record['geneset'],\n",
    "                  'Total target genes': record['genes'],\n",
    "                  'Pathway': record['pathway'],\n",
    "                  'Total pathways': record['pathways']\n",
    "                 })\n",
    "    \n",
    "res_df = pd.DataFrame(out_l)\n",
    "print(res_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway</th>\n",
       "      <th>TF name</th>\n",
       "      <th>TF symbol</th>\n",
       "      <th>Target geneset</th>\n",
       "      <th>Total pathways</th>\n",
       "      <th>Total target genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[nucleus, regulation of transcription from RNA polymerase II promoter, plasma membrane, endoplasmic reticulum membrane, Transmembrane transport of small molecules, Metabolism, integral component of membrane, response to drug, negative regulation of apoptotic process, cellular homeostasis, endoplasmic reticulum, neuron projection, axon, integral component of plasma membrane, cellular response to retinoic acid, extracellular exosome, dendrite, glycoprotein catabolic process, Protein processing in endoplasmic reticulum, protein complex, nuclear membrane, positive regulation of cell migration, apical plasma membrane, cell-cell junction, cell surface, Post-translational protein modification, Metabolism of proteins, response to calcium ion, brush border, cellular response to oxygen-glucose deprivation, protein homooligomerization, positive regulation of transcription from RNA polymerase II promoter, Renin secretion, positive regulation of transcription, DNA-templated, transcription from RNA polymerase II promoter, protein ubiquitination, N-glycan trimming in the ER and Calnexin/Calreticulin cycle, Asparagine N-linked glycosylation, cellular response to hydrogen peroxide, cellular response to oxidative stress, positive regulation of angiogenesis, Vasopressin regulates renal water homeostasis via Aquaporins, Aquaporin-mediated transport, Vasopressin-regulated water reabsorption, renal water homeostasis, positive regulation of epithelial cell migration, caveola, erythrocyte differentiation, response to retinoic acid, basolateral plasma membrane, wound healing, protein folding, negative regulation of cysteine-type endopeptidase activity involved in apoptotic process, brush border membrane, inflammatory response, cellular response to mechanical stimulus, transmembrane transport, apical part of cell, cellular response to hypoxia, positive regulation of fibroblast proliferation, sarcolemma, O2/CO2 exchange in erythrocytes, bicarbonate transport, Erythrocytes take up carbon dioxide and release oxygen, cellular response to UV, establishment or maintenance of actin cytoskeleton polarity, anatomical structure morphogenesis, axon terminus, Proximal tubule bicarbonate reclamation, transport, odontogenesis, metanephric glomerulus vasculature development, secretion by cell, potassium ion transport, potassium ion transmembrane transport, glomerular filtration, lateral ventricle development, sensory perception of pain, cellular response to dexamethasone stimulus, cellular response to cAMP]</td>\n",
       "      <td>Sp1 transcription factor</td>\n",
       "      <td>SP1</td>\n",
       "      <td>[RBBP8, SGK1, SERINC3, ATG5, TEX2, UNC13B, TAOK3, STMN2, ABCA1, IL6ST, STT3B, XPO1, ADAM17, TCEAL1, PHTF1, RPAP2, IGF1R, PIGK, NUCB2, TRPC1, BCL2L13, RAB22A, PTPRA, DNM1L, JKAMP, EIF2AK3, BACE2, HSPH1, CNOT7, PPP3CB, PNRC1, SDCCAG8, CALCOCO1, SRP54, TPR, ZEB2, SLC35F5, LYPD3, MAFB, RNF139, MLLT3, ARL6IP5, ATL1, PKN2, ARIH1, SETX, ZFP91, PARP8, OSBPL9, ANXA4, GPD1L, BTG1, TNFSF10, ARID4A, UBA6, XRN2, LRIG1, RAB11A, JAK2, ZNF148, SMARCA5, VPS29, FBXO11, CCND1, ADK, IGFBP2, SQOR, RAB10, APPBP2, LEPR, TNC, ELF1, CRYAB, ETFDH, GNA13, UBE4B, TRPM7, GPM6B, CROT, MGAT4B, CLDND1, PAPOLG, FAS, HK2, MOCS2, MLPH, TIMP2, ARL1, ASTN1, PCNP, CASK, PPFIA2, MITF, UST, HAT1, CIT, KDR, FEM1C, ATG3, MT2A, ...]</td>\n",
       "      <td>80</td>\n",
       "      <td>402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[integral component of plasma membrane, integral component of membrane, endoplasmic reticulum, plasma membrane, neuron projection, axon, extracellular exosome, sensory perception of pain, nucleus, protein ubiquitination, Post-translational protein modification, Metabolism of proteins, Asparagine N-linked glycosylation, positive regulation of cell migration, negative regulation of apoptotic process, endoplasmic reticulum membrane, Metabolism, Protein processing in endoplasmic reticulum, transport, protein complex, positive regulation of transcription from RNA polymerase II promoter, transmembrane transport, transcription from RNA polymerase II promoter, cell-cell junction, nuclear membrane, positive regulation of transcription, DNA-templated, Vasopressin regulates renal water homeostasis via Aquaporins, Aquaporin-mediated transport, Vasopressin-regulated water reabsorption, Transmembrane transport of small molecules, renal water homeostasis, positive regulation of epithelial cell migration, regulation of keratinocyte differentiation, cell surface, response to drug, response to calcium ion, regulation of transcription from RNA polymerase II promoter, cellular response to dexamethasone stimulus, excretion, water homeostasis, brush border membrane, brush border, basal plasma membrane, apical plasma membrane, cellular response to hypoxia, cellular response to cAMP, response to vitamin D, basolateral plasma membrane, dendrite, apical part of cell, cellular response to hydrogen peroxide, cellular response to UV, potassium ion transport, wound healing, sarcolemma, protein homooligomerization, negative regulation of cysteine-type endopeptidase activity involved in apoptotic process, anatomical structure morphogenesis, cellular response to oxidative stress, axon terminus, Renin secretion, inflammatory response, positive regulation of fibroblast proliferation, cellular response to mechanical stimulus, response to retinoic acid, cellular response to retinoic acid, odontogenesis, metanephric glomerulus vasculature development, erythrocyte differentiation, pancreatic juice secretion, lateral ventricle development, potassium ion transmembrane transport, Proximal tubule bicarbonate reclamation, N-glycan trimming in the ER and Calnexin/Calreticulin cycle, protein folding, positive regulation of angiogenesis, caveola, glycoprotein catabolic process, protein deglycosylation, Bile secretion, O2/CO2 exchange in erythrocytes, bicarbonate transport, glycerol transport, carbon dioxide transport, Erythrocytes take up carbon dioxide and release oxygen, Erythrocytes take up oxygen and release carbon dioxide, cell volume homeostasis, water transport, ammonium transport, Passive transport by Aquaporins, carbon dioxide transmembrane transport, multicellular organismal water homeostasis, cellular response to copper ion, cellular hyperosmotic response, cellular response to nitric oxide, ammonium transmembrane transport, maintenance of symbiont-containing vacuole by host, cellular homeostasis, nitric oxide transport, cellular response to salt stress, ...]</td>\n",
       "      <td>lymphoid enhancer binding factor 1</td>\n",
       "      <td>LEF1</td>\n",
       "      <td>[GPR137B, TEX2, TAOK3, STMN2, RAP2A, TMEM62, TMEM35A, DLG2, SF3B1, UBR3, JDP2, PHTF1, WDR11, GOLGB1, IGF1R, SEC24B, RAB22A, GRIK2, NAALADL2, CRLS1, HNRNPLL, PLOD2, HSPH1, CNOT7, ZNF664, PNRC1, SLC22A23, PBX3, SDCCAG8, RB1CC1, ZEB2, IPO11, TXLNG, MAFB, ZNF569, PKN2, ARIH1, AKT3, TBL1XR1, ARRDC3, ZFP91, PARP8, FNBP1L, HSF2, TNFSF10, XRN2, RAB11A, ROCK2, ZNF148, FBXO11, SYNE2, AK9, NOTCH2, CCND1, LAMA4, LRCH2, CHD1, FBXO32, CSNK1G3, RAB10, STXBP6, PCNX1, NEDD4L, RANBP9, ELF1, GNA13, PDS5A, ACTR3, UBE4B, CTSC, GPM6B, LAMB2, DST, USP34, STC1, KREMEN1, SWT1, ASTN1, AFDN, TOB1, CASK, BCAR3, URI1, COMMD3, SVIL, CHMP2B, MITF, UST, RHOQ, VPS50, FEM1C, MT2A, ZNF644, SLC44A1, BNIP2, UBE3A, MFSD14A, NCAM1, FAP, EHBP1, ...]</td>\n",
       "      <td>111</td>\n",
       "      <td>340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[integral component of membrane, nucleus, inflammatory response, Metabolism of proteins, plasma membrane, extracellular exosome, Renin secretion, response to drug, Metabolism, apical plasma membrane, Vasopressin regulates renal water homeostasis via Aquaporins, Aquaporin-mediated transport, Vasopressin-regulated water reabsorption, dendrite, Transmembrane transport of small molecules, Bile secretion, negative regulation of apoptotic process, cell-cell junction, protein complex, positive regulation of transcription from RNA polymerase II promoter, positive regulation of transcription, DNA-templated, integral component of plasma membrane, transport, caveola, positive regulation of cell migration, erythrocyte differentiation, regulation of transcription from RNA polymerase II promoter, neuron projection, brush border membrane, Post-translational protein modification, transcription from RNA polymerase II promoter, brush border, axon, protein ubiquitination, anatomical structure morphogenesis, cellular response to retinoic acid, positive regulation of fibroblast proliferation, nuclear membrane, cell surface, endoplasmic reticulum membrane, Asparagine N-linked glycosylation, Protein processing in endoplasmic reticulum, response to calcium ion, cellular response to hypoxia, cellular response to cAMP, response to vitamin D, endoplasmic reticulum, protein folding, regulation of keratinocyte differentiation, protein homooligomerization, sensory perception of pain, cellular response to dexamethasone stimulus, wound healing, cellular response to UV, basal plasma membrane, Proximal tubule bicarbonate reclamation, potassium ion transport, cellular response to mechanical stimulus, O2/CO2 exchange in erythrocytes, bicarbonate transport, glycerol transport, carbon dioxide transport, positive regulation of angiogenesis, basolateral plasma membrane, negative regulation of cysteine-type endopeptidase activity involved in apoptotic process, cellular response to hydrogen peroxide, Erythrocytes take up carbon dioxide and release oxygen, Erythrocytes take up oxygen and release carbon dioxide, cell volume homeostasis, potassium ion transmembrane transport, odontogenesis, water transport, apical part of cell, renal water homeostasis, ammonium transport, Passive transport by Aquaporins, sarcolemma, carbon dioxide transmembrane transport, pancreatic juice secretion, multicellular organismal water homeostasis, cellular response to copper ion, cellular hyperosmotic response, cellular response to nitric oxide, ammonium transmembrane transport, maintenance of symbiont-containing vacuole by host, cellular homeostasis, nitric oxide transport, cellular response to salt stress, cellular response to mercury ion, renal water transport, transepithelial water transport, cGMP biosynthetic process, cellular response to stress, symbiont-containing vacuole membrane, lateral ventricle development, positive regulation of saliva secretion, cerebrospinal fluid secretion, establishment or maintenance of actin cytoskeleton polarity, symbiont-containing vacuole, cellular response to inorganic substance, ...]</td>\n",
       "      <td>MYC associated zinc finger protein</td>\n",
       "      <td>MAZ</td>\n",
       "      <td>[CYYR1, PGRMC2, PAPOLG, LOXL3, XRN2, STXBP6, EIF4A1, GNAS, DPT, CDC42EP5, CFL1, TINAGL1, PNRC1, RASSF2, NCOA3, SLC6A9, JAK2, EHBP1, CHD1, SHANK2, FBXL19, CSDC2, ID4, EBF1, ADAMTS4, ACTC1, PKN3, DOCK4, MNT, ACTR3, PHF20L1, STMN2, TRIM41, IGF2BP3, MITF, WNT2, SYNE1, ADAMTS15, BCAM, TGIF2, ZNF703, PPM1A, EFNB1, SMC6, PARP12, PTMA, EPB41L1, KDELR1, ADAM19, UBA3, LRFN5, ATF6B, CCND1, LAMA5, SUMO1, RB1CC1, LRRC4, XPO1, STC1, CUL2, GRN, ERF, RGS7, TCF21, HNRNPA3, WNK4, STAG2, PELP1, CD109, HOXA13, SDCCAG8, BCAR3, TMEM35A, ACAN, ZNF644, ATP5MC2, ASIC1, CD47, RAB10, ATL1, MARCH7, CBX6, EIF5A, SWT1, HNRNPA2B1, DLG2, CCP110, FSTL3, FGFRL1, EFNB3, GDNF, UST, CBX3, UBE2D3, HNRNPA0, MMP2, TFAP2C, GPM6B, TLK1, WAC, ...]</td>\n",
       "      <td>106</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[nucleus, protein complex, plasma membrane, positive regulation of transcription from RNA polymerase II promoter, keratan sulfate biosynthetic process, extracellular exosome, Metabolism, neuron projection, axon, integral component of membrane, sensory perception of pain, protein ubiquitination, Renin secretion, response to drug, apical plasma membrane, Vasopressin regulates renal water homeostasis via Aquaporins, Aquaporin-mediated transport, Vasopressin-regulated water reabsorption, dendrite, Transmembrane transport of small molecules, Bile secretion, integral component of plasma membrane, positive regulation of cell migration, negative regulation of apoptotic process, Post-translational protein modification, Metabolism of proteins, endoplasmic reticulum membrane, Protein processing in endoplasmic reticulum, endoplasmic reticulum, protein homooligomerization, transcription from RNA polymerase II promoter, nuclear membrane, positive regulation of transcription, DNA-templated, cellular response to dexamethasone stimulus, excretion, water homeostasis, brush border membrane, cellular response to hypoxia, Asparagine N-linked glycosylation, cell surface, basal plasma membrane, cellular response to cAMP, response to vitamin D, transmembrane transport, cellular response to UV, regulation of transcription from RNA polymerase II promoter, camera-type eye morphogenesis, protein folding, O2/CO2 exchange in erythrocytes, bicarbonate transport, Erythrocytes take up carbon dioxide and release oxygen, cell-cell junction, positive regulation of angiogenesis, positive regulation of epithelial cell migration, wound healing, secretion by cell, secretory granule organization, axon terminus, apical part of cell, negative regulation of cysteine-type endopeptidase activity involved in apoptotic process, cellular response to retinoic acid, odontogenesis, metanephric glomerulus vasculature development, inflammatory response, cellular response to hydrogen peroxide, pancreatic juice secretion, transport, cellular response to mechanical stimulus, glomerular filtration, anatomical structure morphogenesis, potassium ion transport, sarcolemma, potassium ion transmembrane transport, basolateral plasma membrane, Proximal tubule bicarbonate reclamation, brush border, glycerol transport, carbon dioxide transport, Erythrocytes take up oxygen and release carbon dioxide, cell volume homeostasis, water transport, renal water homeostasis, ammonium transport, Passive transport by Aquaporins, positive regulation of fibroblast proliferation, carbon dioxide transmembrane transport, multicellular organismal water homeostasis, cellular response to copper ion, cellular hyperosmotic response, cellular response to nitric oxide, ammonium transmembrane transport, maintenance of symbiont-containing vacuole by host, cellular homeostasis, nitric oxide transport, cellular response to salt stress, cellular response to mercury ion, renal water transport, transepithelial water transport, cGMP biosynthetic process, cellular response to stress, ...]</td>\n",
       "      <td>transcription regulator</td>\n",
       "      <td>E12</td>\n",
       "      <td>[STK3, UNC13B, EGLN1, LUM, STMN2, TMEM62, TMEM35A, FNDC3A, DLG2, UBR3, CASC4, GNAS, IGF1R, RAB22A, ACSL4, CRLS1, EIF2AK3, PLOD2, SPINT2, ZNF664, ZEB2, TXLNG, MLLT3, ARL6IP5, PKN2, ARRDC3, ZFP91, PTGFR, FNBP1L, OSBPL9, GPD1L, LRIG1, GOLGA4, CDK14, SYNE2, AK9, LAMA4, ADK, LTBP2, FBXO32, NEDD4L, RANBP9, ZFR, GNA13, ACAA2, GPM6B, MMP15, MGAT4B, CLDND1, DST, PAPOLG, CMTM4, ZNF189, STC1, CHMP2B, UST, RHOQ, ZNF644, USP1, SDK1, SLC44A1, POLK, NCAM1, ADGRB2, KIF5B, RTN3, DSTN, KTN1, KIF13A, UCHL3, ITGA7, SYNE1, TFAP2C, AGL, COL8A1, DUSP3, TAB2, RSF1, TIGAR, RGS7, SCAMP5, USO1, DENND4A, SKIL, CYB5R4, PIK3R1, HIF1A, IDE, DGKA, ELK3, RAP1GAP2, TSC22D1, JMJD1C, JAZF1, PHKB, HMCN1, CYLD, ENPP5, C1GALT1C1, EPB41, ...]</td>\n",
       "      <td>107</td>\n",
       "      <td>238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[integral component of plasma membrane, plasma membrane, nucleus, endoplasmic reticulum membrane, Transmembrane transport of small molecules, Metabolism, integral component of membrane, endoplasmic reticulum, regulation of transcription from RNA polymerase II promoter, neuron projection, axon, extracellular exosome, negative regulation of apoptotic process, dendrite, sensory perception of pain, Renin secretion, positive regulation of transcription from RNA polymerase II promoter, sarcolemma, response to drug, apical plasma membrane, Vasopressin regulates renal water homeostasis via Aquaporins, Aquaporin-mediated transport, Vasopressin-regulated water reabsorption, Bile secretion, Post-translational protein modification, Metabolism of proteins, Asparagine N-linked glycosylation, positive regulation of cell migration, Protein processing in endoplasmic reticulum, transport, protein homooligomerization, positive regulation of transcription, DNA-templated, transcription from RNA polymerase II promoter, nuclear membrane, protein complex, positive regulation of angiogenesis, protein ubiquitination, cellular response to dexamethasone stimulus, bicarbonate transport, basolateral plasma membrane, basal plasma membrane, cellular response to hypoxia, cellular response to cAMP, response to vitamin D, caveola, cell surface, transmembrane transport, inflammatory response, cellular response to hydrogen peroxide, cellular response to UV, camera-type eye morphogenesis, potassium ion transport, wound healing, positive regulation of fibroblast proliferation, cell-cell junction, positive regulation of epithelial cell migration, cellular response to oxidative stress, erythrocyte differentiation, odontogenesis, cellular response to retinoic acid, axon terminus, potassium ion transmembrane transport, brush border, apical part of cell, protein folding, anatomical structure morphogenesis, response to retinoic acid]</td>\n",
       "      <td>forkhead box O4</td>\n",
       "      <td>FOXO4</td>\n",
       "      <td>[GPR137B, SGK1, SERINC3, TEX2, ZNF770, STMN2, IL6ST, DLG2, PPP3R1, KANK1, GNAS, GOLGB1, IFIH1, IGF1R, SEC24B, EIF3E, ACSL4, GRIK2, NAALADL2, EIF2AK3, HNRNPLL, ZNF664, PPP3CB, PNRC1, PBX3, TPR, ZEB2, TXLNG, MLLT3, PKN2, AKT3, TMEM167A, TBL1XR1, ARRDC3, BTG1, TNFSF10, ARID4A, ZZZ3, ZNF148, FBXO11, LAMA4, BBS7, LRCH2, CHD1, FBXO32, ACKR4, CSNK1G3, HIP1R, RAB10, APPBP2, STXBP6, PCNX1, RANBP9, ELF1, SLC4A7, UBE4B, SAT1, MGAT4B, DST, STC1, KREMEN1, SWT1, TOB1, COMMD3, CHMP2B, STARD5, MITF, SUMO1, HEBP2, RHOQ, DNAJB14, ADTRP, ZNF644, SDK1, SLC44A1, SLMAP, UBE3A, MGST3, SLC30A9, NCAM1, EHBP1, KBTBD2, GNAQ, SATB1, ANXA1, ZBTB20, NDFIP1, KIF2A, KDM3A, ZNF277, FAR1, STXBP3, RAB6A, NAP1L1, ZC3H7A, PCNA, AGL, COL8A1, KCNJ15, STAG2, ...]</td>\n",
       "      <td>67</td>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Pathway  \\\n",
       "0  [nucleus, regulation of transcription from RNA polymerase II promoter, plasma membrane, endoplasmic reticulum membrane, Transmembrane transport of small molecules, Metabolism, integral component of membrane, response to drug, negative regulation of apoptotic process, cellular homeostasis, endoplasmic reticulum, neuron projection, axon, integral component of plasma membrane, cellular response to retinoic acid, extracellular exosome, dendrite, glycoprotein catabolic process, Protein processing in endoplasmic reticulum, protein complex, nuclear membrane, positive regulation of cell migration, apical plasma membrane, cell-cell junction, cell surface, Post-translational protein modification, Metabolism of proteins, response to calcium ion, brush border, cellular response to oxygen-glucose deprivation, protein homooligomerization, positive regulation of transcription from RNA polymerase II promoter, Renin secretion, positive regulation of transcription, DNA-templated, transcription from RNA polymerase II promoter, protein ubiquitination, N-glycan trimming in the ER and Calnexin/Calreticulin cycle, Asparagine N-linked glycosylation, cellular response to hydrogen peroxide, cellular response to oxidative stress, positive regulation of angiogenesis, Vasopressin regulates renal water homeostasis via Aquaporins, Aquaporin-mediated transport, Vasopressin-regulated water reabsorption, renal water homeostasis, positive regulation of epithelial cell migration, caveola, erythrocyte differentiation, response to retinoic acid, basolateral plasma membrane, wound healing, protein folding, negative regulation of cysteine-type endopeptidase activity involved in apoptotic process, brush border membrane, inflammatory response, cellular response to mechanical stimulus, transmembrane transport, apical part of cell, cellular response to hypoxia, positive regulation of fibroblast proliferation, sarcolemma, O2/CO2 exchange in erythrocytes, bicarbonate transport, Erythrocytes take up carbon dioxide and release oxygen, cellular response to UV, establishment or maintenance of actin cytoskeleton polarity, anatomical structure morphogenesis, axon terminus, Proximal tubule bicarbonate reclamation, transport, odontogenesis, metanephric glomerulus vasculature development, secretion by cell, potassium ion transport, potassium ion transmembrane transport, glomerular filtration, lateral ventricle development, sensory perception of pain, cellular response to dexamethasone stimulus, cellular response to cAMP]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
       "1  [integral component of plasma membrane, integral component of membrane, endoplasmic reticulum, plasma membrane, neuron projection, axon, extracellular exosome, sensory perception of pain, nucleus, protein ubiquitination, Post-translational protein modification, Metabolism of proteins, Asparagine N-linked glycosylation, positive regulation of cell migration, negative regulation of apoptotic process, endoplasmic reticulum membrane, Metabolism, Protein processing in endoplasmic reticulum, transport, protein complex, positive regulation of transcription from RNA polymerase II promoter, transmembrane transport, transcription from RNA polymerase II promoter, cell-cell junction, nuclear membrane, positive regulation of transcription, DNA-templated, Vasopressin regulates renal water homeostasis via Aquaporins, Aquaporin-mediated transport, Vasopressin-regulated water reabsorption, Transmembrane transport of small molecules, renal water homeostasis, positive regulation of epithelial cell migration, regulation of keratinocyte differentiation, cell surface, response to drug, response to calcium ion, regulation of transcription from RNA polymerase II promoter, cellular response to dexamethasone stimulus, excretion, water homeostasis, brush border membrane, brush border, basal plasma membrane, apical plasma membrane, cellular response to hypoxia, cellular response to cAMP, response to vitamin D, basolateral plasma membrane, dendrite, apical part of cell, cellular response to hydrogen peroxide, cellular response to UV, potassium ion transport, wound healing, sarcolemma, protein homooligomerization, negative regulation of cysteine-type endopeptidase activity involved in apoptotic process, anatomical structure morphogenesis, cellular response to oxidative stress, axon terminus, Renin secretion, inflammatory response, positive regulation of fibroblast proliferation, cellular response to mechanical stimulus, response to retinoic acid, cellular response to retinoic acid, odontogenesis, metanephric glomerulus vasculature development, erythrocyte differentiation, pancreatic juice secretion, lateral ventricle development, potassium ion transmembrane transport, Proximal tubule bicarbonate reclamation, N-glycan trimming in the ER and Calnexin/Calreticulin cycle, protein folding, positive regulation of angiogenesis, caveola, glycoprotein catabolic process, protein deglycosylation, Bile secretion, O2/CO2 exchange in erythrocytes, bicarbonate transport, glycerol transport, carbon dioxide transport, Erythrocytes take up carbon dioxide and release oxygen, Erythrocytes take up oxygen and release carbon dioxide, cell volume homeostasis, water transport, ammonium transport, Passive transport by Aquaporins, carbon dioxide transmembrane transport, multicellular organismal water homeostasis, cellular response to copper ion, cellular hyperosmotic response, cellular response to nitric oxide, ammonium transmembrane transport, maintenance of symbiont-containing vacuole by host, cellular homeostasis, nitric oxide transport, cellular response to salt stress, ...]                                       \n",
       "2  [integral component of membrane, nucleus, inflammatory response, Metabolism of proteins, plasma membrane, extracellular exosome, Renin secretion, response to drug, Metabolism, apical plasma membrane, Vasopressin regulates renal water homeostasis via Aquaporins, Aquaporin-mediated transport, Vasopressin-regulated water reabsorption, dendrite, Transmembrane transport of small molecules, Bile secretion, negative regulation of apoptotic process, cell-cell junction, protein complex, positive regulation of transcription from RNA polymerase II promoter, positive regulation of transcription, DNA-templated, integral component of plasma membrane, transport, caveola, positive regulation of cell migration, erythrocyte differentiation, regulation of transcription from RNA polymerase II promoter, neuron projection, brush border membrane, Post-translational protein modification, transcription from RNA polymerase II promoter, brush border, axon, protein ubiquitination, anatomical structure morphogenesis, cellular response to retinoic acid, positive regulation of fibroblast proliferation, nuclear membrane, cell surface, endoplasmic reticulum membrane, Asparagine N-linked glycosylation, Protein processing in endoplasmic reticulum, response to calcium ion, cellular response to hypoxia, cellular response to cAMP, response to vitamin D, endoplasmic reticulum, protein folding, regulation of keratinocyte differentiation, protein homooligomerization, sensory perception of pain, cellular response to dexamethasone stimulus, wound healing, cellular response to UV, basal plasma membrane, Proximal tubule bicarbonate reclamation, potassium ion transport, cellular response to mechanical stimulus, O2/CO2 exchange in erythrocytes, bicarbonate transport, glycerol transport, carbon dioxide transport, positive regulation of angiogenesis, basolateral plasma membrane, negative regulation of cysteine-type endopeptidase activity involved in apoptotic process, cellular response to hydrogen peroxide, Erythrocytes take up carbon dioxide and release oxygen, Erythrocytes take up oxygen and release carbon dioxide, cell volume homeostasis, potassium ion transmembrane transport, odontogenesis, water transport, apical part of cell, renal water homeostasis, ammonium transport, Passive transport by Aquaporins, sarcolemma, carbon dioxide transmembrane transport, pancreatic juice secretion, multicellular organismal water homeostasis, cellular response to copper ion, cellular hyperosmotic response, cellular response to nitric oxide, ammonium transmembrane transport, maintenance of symbiont-containing vacuole by host, cellular homeostasis, nitric oxide transport, cellular response to salt stress, cellular response to mercury ion, renal water transport, transepithelial water transport, cGMP biosynthetic process, cellular response to stress, symbiont-containing vacuole membrane, lateral ventricle development, positive regulation of saliva secretion, cerebrospinal fluid secretion, establishment or maintenance of actin cytoskeleton polarity, symbiont-containing vacuole, cellular response to inorganic substance, ...]   \n",
       "3  [nucleus, protein complex, plasma membrane, positive regulation of transcription from RNA polymerase II promoter, keratan sulfate biosynthetic process, extracellular exosome, Metabolism, neuron projection, axon, integral component of membrane, sensory perception of pain, protein ubiquitination, Renin secretion, response to drug, apical plasma membrane, Vasopressin regulates renal water homeostasis via Aquaporins, Aquaporin-mediated transport, Vasopressin-regulated water reabsorption, dendrite, Transmembrane transport of small molecules, Bile secretion, integral component of plasma membrane, positive regulation of cell migration, negative regulation of apoptotic process, Post-translational protein modification, Metabolism of proteins, endoplasmic reticulum membrane, Protein processing in endoplasmic reticulum, endoplasmic reticulum, protein homooligomerization, transcription from RNA polymerase II promoter, nuclear membrane, positive regulation of transcription, DNA-templated, cellular response to dexamethasone stimulus, excretion, water homeostasis, brush border membrane, cellular response to hypoxia, Asparagine N-linked glycosylation, cell surface, basal plasma membrane, cellular response to cAMP, response to vitamin D, transmembrane transport, cellular response to UV, regulation of transcription from RNA polymerase II promoter, camera-type eye morphogenesis, protein folding, O2/CO2 exchange in erythrocytes, bicarbonate transport, Erythrocytes take up carbon dioxide and release oxygen, cell-cell junction, positive regulation of angiogenesis, positive regulation of epithelial cell migration, wound healing, secretion by cell, secretory granule organization, axon terminus, apical part of cell, negative regulation of cysteine-type endopeptidase activity involved in apoptotic process, cellular response to retinoic acid, odontogenesis, metanephric glomerulus vasculature development, inflammatory response, cellular response to hydrogen peroxide, pancreatic juice secretion, transport, cellular response to mechanical stimulus, glomerular filtration, anatomical structure morphogenesis, potassium ion transport, sarcolemma, potassium ion transmembrane transport, basolateral plasma membrane, Proximal tubule bicarbonate reclamation, brush border, glycerol transport, carbon dioxide transport, Erythrocytes take up oxygen and release carbon dioxide, cell volume homeostasis, water transport, renal water homeostasis, ammonium transport, Passive transport by Aquaporins, positive regulation of fibroblast proliferation, carbon dioxide transmembrane transport, multicellular organismal water homeostasis, cellular response to copper ion, cellular hyperosmotic response, cellular response to nitric oxide, ammonium transmembrane transport, maintenance of symbiont-containing vacuole by host, cellular homeostasis, nitric oxide transport, cellular response to salt stress, cellular response to mercury ion, renal water transport, transepithelial water transport, cGMP biosynthetic process, cellular response to stress, ...]                                                                        \n",
       "4  [integral component of plasma membrane, plasma membrane, nucleus, endoplasmic reticulum membrane, Transmembrane transport of small molecules, Metabolism, integral component of membrane, endoplasmic reticulum, regulation of transcription from RNA polymerase II promoter, neuron projection, axon, extracellular exosome, negative regulation of apoptotic process, dendrite, sensory perception of pain, Renin secretion, positive regulation of transcription from RNA polymerase II promoter, sarcolemma, response to drug, apical plasma membrane, Vasopressin regulates renal water homeostasis via Aquaporins, Aquaporin-mediated transport, Vasopressin-regulated water reabsorption, Bile secretion, Post-translational protein modification, Metabolism of proteins, Asparagine N-linked glycosylation, positive regulation of cell migration, Protein processing in endoplasmic reticulum, transport, protein homooligomerization, positive regulation of transcription, DNA-templated, transcription from RNA polymerase II promoter, nuclear membrane, protein complex, positive regulation of angiogenesis, protein ubiquitination, cellular response to dexamethasone stimulus, bicarbonate transport, basolateral plasma membrane, basal plasma membrane, cellular response to hypoxia, cellular response to cAMP, response to vitamin D, caveola, cell surface, transmembrane transport, inflammatory response, cellular response to hydrogen peroxide, cellular response to UV, camera-type eye morphogenesis, potassium ion transport, wound healing, positive regulation of fibroblast proliferation, cell-cell junction, positive regulation of epithelial cell migration, cellular response to oxidative stress, erythrocyte differentiation, odontogenesis, cellular response to retinoic acid, axon terminus, potassium ion transmembrane transport, brush border, apical part of cell, protein folding, anatomical structure morphogenesis, response to retinoic acid]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
       "\n",
       "                              TF name TF symbol  \\\n",
       "0  Sp1 transcription factor            SP1        \n",
       "1  lymphoid enhancer binding factor 1  LEF1       \n",
       "2  MYC associated zinc finger protein  MAZ        \n",
       "3  transcription regulator             E12        \n",
       "4  forkhead box O4                     FOXO4      \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  Target geneset  \\\n",
       "0  [RBBP8, SGK1, SERINC3, ATG5, TEX2, UNC13B, TAOK3, STMN2, ABCA1, IL6ST, STT3B, XPO1, ADAM17, TCEAL1, PHTF1, RPAP2, IGF1R, PIGK, NUCB2, TRPC1, BCL2L13, RAB22A, PTPRA, DNM1L, JKAMP, EIF2AK3, BACE2, HSPH1, CNOT7, PPP3CB, PNRC1, SDCCAG8, CALCOCO1, SRP54, TPR, ZEB2, SLC35F5, LYPD3, MAFB, RNF139, MLLT3, ARL6IP5, ATL1, PKN2, ARIH1, SETX, ZFP91, PARP8, OSBPL9, ANXA4, GPD1L, BTG1, TNFSF10, ARID4A, UBA6, XRN2, LRIG1, RAB11A, JAK2, ZNF148, SMARCA5, VPS29, FBXO11, CCND1, ADK, IGFBP2, SQOR, RAB10, APPBP2, LEPR, TNC, ELF1, CRYAB, ETFDH, GNA13, UBE4B, TRPM7, GPM6B, CROT, MGAT4B, CLDND1, PAPOLG, FAS, HK2, MOCS2, MLPH, TIMP2, ARL1, ASTN1, PCNP, CASK, PPFIA2, MITF, UST, HAT1, CIT, KDR, FEM1C, ATG3, MT2A, ...]                                     \n",
       "1  [GPR137B, TEX2, TAOK3, STMN2, RAP2A, TMEM62, TMEM35A, DLG2, SF3B1, UBR3, JDP2, PHTF1, WDR11, GOLGB1, IGF1R, SEC24B, RAB22A, GRIK2, NAALADL2, CRLS1, HNRNPLL, PLOD2, HSPH1, CNOT7, ZNF664, PNRC1, SLC22A23, PBX3, SDCCAG8, RB1CC1, ZEB2, IPO11, TXLNG, MAFB, ZNF569, PKN2, ARIH1, AKT3, TBL1XR1, ARRDC3, ZFP91, PARP8, FNBP1L, HSF2, TNFSF10, XRN2, RAB11A, ROCK2, ZNF148, FBXO11, SYNE2, AK9, NOTCH2, CCND1, LAMA4, LRCH2, CHD1, FBXO32, CSNK1G3, RAB10, STXBP6, PCNX1, NEDD4L, RANBP9, ELF1, GNA13, PDS5A, ACTR3, UBE4B, CTSC, GPM6B, LAMB2, DST, USP34, STC1, KREMEN1, SWT1, ASTN1, AFDN, TOB1, CASK, BCAR3, URI1, COMMD3, SVIL, CHMP2B, MITF, UST, RHOQ, VPS50, FEM1C, MT2A, ZNF644, SLC44A1, BNIP2, UBE3A, MFSD14A, NCAM1, FAP, EHBP1, ...]                 \n",
       "2  [CYYR1, PGRMC2, PAPOLG, LOXL3, XRN2, STXBP6, EIF4A1, GNAS, DPT, CDC42EP5, CFL1, TINAGL1, PNRC1, RASSF2, NCOA3, SLC6A9, JAK2, EHBP1, CHD1, SHANK2, FBXL19, CSDC2, ID4, EBF1, ADAMTS4, ACTC1, PKN3, DOCK4, MNT, ACTR3, PHF20L1, STMN2, TRIM41, IGF2BP3, MITF, WNT2, SYNE1, ADAMTS15, BCAM, TGIF2, ZNF703, PPM1A, EFNB1, SMC6, PARP12, PTMA, EPB41L1, KDELR1, ADAM19, UBA3, LRFN5, ATF6B, CCND1, LAMA5, SUMO1, RB1CC1, LRRC4, XPO1, STC1, CUL2, GRN, ERF, RGS7, TCF21, HNRNPA3, WNK4, STAG2, PELP1, CD109, HOXA13, SDCCAG8, BCAR3, TMEM35A, ACAN, ZNF644, ATP5MC2, ASIC1, CD47, RAB10, ATL1, MARCH7, CBX6, EIF5A, SWT1, HNRNPA2B1, DLG2, CCP110, FSTL3, FGFRL1, EFNB3, GDNF, UST, CBX3, UBE2D3, HNRNPA0, MMP2, TFAP2C, GPM6B, TLK1, WAC, ...]                      \n",
       "3  [STK3, UNC13B, EGLN1, LUM, STMN2, TMEM62, TMEM35A, FNDC3A, DLG2, UBR3, CASC4, GNAS, IGF1R, RAB22A, ACSL4, CRLS1, EIF2AK3, PLOD2, SPINT2, ZNF664, ZEB2, TXLNG, MLLT3, ARL6IP5, PKN2, ARRDC3, ZFP91, PTGFR, FNBP1L, OSBPL9, GPD1L, LRIG1, GOLGA4, CDK14, SYNE2, AK9, LAMA4, ADK, LTBP2, FBXO32, NEDD4L, RANBP9, ZFR, GNA13, ACAA2, GPM6B, MMP15, MGAT4B, CLDND1, DST, PAPOLG, CMTM4, ZNF189, STC1, CHMP2B, UST, RHOQ, ZNF644, USP1, SDK1, SLC44A1, POLK, NCAM1, ADGRB2, KIF5B, RTN3, DSTN, KTN1, KIF13A, UCHL3, ITGA7, SYNE1, TFAP2C, AGL, COL8A1, DUSP3, TAB2, RSF1, TIGAR, RGS7, SCAMP5, USO1, DENND4A, SKIL, CYB5R4, PIK3R1, HIF1A, IDE, DGKA, ELK3, RAP1GAP2, TSC22D1, JMJD1C, JAZF1, PHKB, HMCN1, CYLD, ENPP5, C1GALT1C1, EPB41, ...]                        \n",
       "4  [GPR137B, SGK1, SERINC3, TEX2, ZNF770, STMN2, IL6ST, DLG2, PPP3R1, KANK1, GNAS, GOLGB1, IFIH1, IGF1R, SEC24B, EIF3E, ACSL4, GRIK2, NAALADL2, EIF2AK3, HNRNPLL, ZNF664, PPP3CB, PNRC1, PBX3, TPR, ZEB2, TXLNG, MLLT3, PKN2, AKT3, TMEM167A, TBL1XR1, ARRDC3, BTG1, TNFSF10, ARID4A, ZZZ3, ZNF148, FBXO11, LAMA4, BBS7, LRCH2, CHD1, FBXO32, ACKR4, CSNK1G3, HIP1R, RAB10, APPBP2, STXBP6, PCNX1, RANBP9, ELF1, SLC4A7, UBE4B, SAT1, MGAT4B, DST, STC1, KREMEN1, SWT1, TOB1, COMMD3, CHMP2B, STARD5, MITF, SUMO1, HEBP2, RHOQ, DNAJB14, ADTRP, ZNF644, SDK1, SLC44A1, SLMAP, UBE3A, MGST3, SLC30A9, NCAM1, EHBP1, KBTBD2, GNAQ, SATB1, ANXA1, ZBTB20, NDFIP1, KIF2A, KDM3A, ZNF277, FAR1, STXBP3, RAB6A, NAP1L1, ZC3H7A, PCNA, AGL, COL8A1, KCNJ15, STAG2, ...]   \n",
       "\n",
       "   Total pathways  Total target genes  \n",
       "0  80              402                 \n",
       "1  111             340                 \n",
       "2  106             273                 \n",
       "3  107             238                 \n",
       "4  67              237                 "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Summary table\n",
    "res_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NETWORK HYPOTHESES"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### template that involves genes and pathways\n",
    "* **Query 1: ( L = 3 ) (ngly1-ppi-pw-aqp1) ngly1-[interacts with]-gene--phys--aqp1 **\n",
    "    * without degree filters\n",
    "    * this query in graph v2 gives 0 paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total paths: 1670\n",
      "CPU times: user 1.5 ms, sys: 142 µs, total: 1.64 ms\n",
      "Wall time: 511 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 1\n",
    "with driver.session() as session:\n",
    "    result = session.run(\n",
    "    \"\"\"\n",
    "    MATCH path=(source:GENE)-[:`RO:0002434`]-(:GENE)--(:PHYS)--(target:GENE)\n",
    "\n",
    "    WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "    WITH path,\n",
    "\n",
    "    [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked,\n",
    "\n",
    "    [r IN relationships(path) WHERE r.property_label IN ['in paralogy relationship with','in orthology relationship with','colocalizes with']] AS edges_marked\n",
    "\n",
    "    WHERE size(nodes_marked) = 0 AND size(edges_marked) = 0\n",
    "\n",
    "    RETURN count(distinct path) as paths\n",
    "    \"\"\"\n",
    "    )\n",
    "\n",
    "# print result\n",
    "for record in result:\n",
    "    print('Total paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### template rna-reg L=3 min\n",
    "* **Query 14: TF-pw (L=4) NGLY1--RNA_GENE--TF--PHYS--AQP1**\n",
    "    * filter expression edge \n",
    "    * filter TF regulatory edge and directionality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 0 ns, sys: 4.37 ms, total: 4.37 ms\n",
      "Wall time: 32.5 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 14\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE {id: 'HGNC:17646'})-[i1:`RO:0002434`]-(rna:GENE)<-[i2:`RO:0002434`]-(tf:GENE)--(pw:PHYS)--(target:GENE {id: 'HGNC:633'})\n",
    "\n",
    "        WHERE ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,i2,rna,tf,pw,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 \n",
    "        \n",
    "        AND toLower(i1.reference_supporting_text) =~ '.*freeze.*' \n",
    "        \n",
    "        AND toLower(i2.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN count(distinct rna) as rnas, count(distinct tf) as tfs, count(distinct pw) as pws, count(distinct path) as paths\n",
    "        \n",
    "        ORDER BY paths DESC\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "out_l = list()\n",
    "for record in result:\n",
    "    out_l.append({'Expressed_genes': record['rnas'], \n",
    "                  'TFs': record['tfs'],\n",
    "                  'Pathways': record['pws'],\n",
    "                  'Paths': record['paths']})\n",
    "    \n",
    "res_df = pd.DataFrame(out_l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Expressed_genes</th>\n",
       "      <th>Paths</th>\n",
       "      <th>Pathways</th>\n",
       "      <th>TFs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1349</td>\n",
       "      <td>13467</td>\n",
       "      <td>29</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Expressed_genes  Paths  Pathways  TFs\n",
       "0  1349             13467  29        221"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Summary table\n",
    "res_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Query 14.1: TF-pw (L=3) NGLY1--RNA_GENE--TF--AQP1**\n",
    "    * filter expression edge \n",
    "    * filter TF regulatory edge and directionality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.96 ms, sys: 0 ns, total: 2.96 ms\n",
      "Wall time: 1.56 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 14.1\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE {id: 'HGNC:17646'})-[i1:`RO:0002434`]-(rna:GENE)<-[i2:`RO:0002434`]-(tf:GENE)--(target:GENE {id: 'HGNC:633'})\n",
    "\n",
    "        WHERE ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,i2,rna,tf,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 \n",
    "        \n",
    "        AND toLower(i1.reference_supporting_text) =~ '.*freeze.*' \n",
    "        \n",
    "        AND toLower(i2.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN count(distinct rna) as rnas, count(distinct tf) as tfs, count(distinct path) as paths\n",
    "        \n",
    "        ORDER BY paths DESC\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "out_l = list()\n",
    "for record in result:\n",
    "    out_l.append({'Expressed_genes': record['rnas'], \n",
    "                  'TFs': record['tfs'],\n",
    "                  'Paths': record['paths']})\n",
    "    \n",
    "res_df = pd.DataFrame(out_l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Expressed_genes</th>\n",
       "      <th>Paths</th>\n",
       "      <th>TFs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>818</td>\n",
       "      <td>1839</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Expressed_genes  Paths  TFs\n",
       "0  818              1839   13 "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# summary table\n",
    "res_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look for a template of interest (only regulation edge): Open query + metapaths \n",
    "* **Query 13: Variable length 2-4 and pattern**\n",
    "    * pattern: ngly1<-regulatory-tf-[<=3]-aqp1 \n",
    "    * ngly1-[interacts with]-gene-[1..3]-aqp1 \n",
    "    * filter specific regulatory types on specific edges "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 11213\n",
      "CPU times: user 1.32 ms, sys: 3.02 ms, total: 4.34 ms\n",
      "Wall time: 52.2 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 13\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)<-[i1:`RO:0002434`]-(tf:GENE)-[*..3]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 AND toLower(i1.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5.17 ms, sys: 3.32 ms, total: 8.48 ms\n",
      "Wall time: 52.6 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 13\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)<-[i1:`RO:0002434`]-(tf:GENE)-[*..3]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 AND toLower(i1.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN DISTINCT extract (x in rels(path) | type(x)) as types, extract (n in nodes(path) | labels(n)) as labels, length(path) as mp_length, count(distinct path) as paths \n",
    "        \n",
    "        ORDER BY mp_length, paths DESC\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "out_l = list()\n",
    "for record in result:\n",
    "    out_l.append({'Nodes': record['labels'], \n",
    "                  'Relations': record['types'],\n",
    "                  'Metapath length': record['mp_length'],\n",
    "                  'Paths': record['paths']})\n",
    "    \n",
    "res_df = pd.DataFrame(out_l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metapath length</th>\n",
       "      <th>Nodes</th>\n",
       "      <th>Paths</th>\n",
       "      <th>Relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>36</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>8</td>\n",
       "      <td>[RO:0002434, RO:HOM0000011, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>[[GENE], [GENE], [ANAT], [GENE]]</td>\n",
       "      <td>5</td>\n",
       "      <td>[RO:0002434, RO:0002206, RO:0002206]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>6534</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002434, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [ANAT], [GENE]]</td>\n",
       "      <td>1502</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002206, RO:0002206]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>432</td>\n",
       "      <td>[RO:0002434, RO:0002434, BFO:0000050, BFO:0000050]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>425</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002331, RO:0002331]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>400</td>\n",
       "      <td>[RO:0002434, RO:HOM0000011, RO:0002434, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>208</td>\n",
       "      <td>[RO:0002434, RO:HOM0000011, RO:HOM0000011, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>190</td>\n",
       "      <td>[RO:0002434, RO:0002325, RO:0002434, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [DISO], [GENE], [GENE]]</td>\n",
       "      <td>163</td>\n",
       "      <td>[RO:0002434, RO:0002200, RO:0002200, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [ANAT], [GENE], [GENE]]</td>\n",
       "      <td>145</td>\n",
       "      <td>[RO:0002434, RO:0002206, RO:0002206, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>136</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002325, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [DISO], [PHYS], [GENE]]</td>\n",
       "      <td>113</td>\n",
       "      <td>[RO:0002434, RO:0002200, None, RO:0002331]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>106</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002327, RO:0002327]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>96</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:HOM0000011, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [ANAT], [GENE]]</td>\n",
       "      <td>94</td>\n",
       "      <td>[RO:0002434, RO:HOM0000011, RO:0002206, RO:0002206]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [PHYS], [GENE], [GENE]]</td>\n",
       "      <td>87</td>\n",
       "      <td>[RO:0002434, RO:0002331, RO:0002331, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>72</td>\n",
       "      <td>[RO:0002434, RO:0002435, RO:0002434, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>64</td>\n",
       "      <td>[RO:0002434, RO:HOM0000017, RO:HOM0000017, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [DISO], [GENE]]</td>\n",
       "      <td>60</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002607, RO:0002607]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>56</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002434, RO:HOM0000011]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [ANAT], [GENE]]</td>\n",
       "      <td>54</td>\n",
       "      <td>[RO:0002434, RO:0002325, RO:0002206, RO:0002206]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [PHYS], [GENE], [GENE]]</td>\n",
       "      <td>35</td>\n",
       "      <td>[RO:0002434, RO:0002327, RO:0002327, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [DISO], [PHYS], [GENE]]</td>\n",
       "      <td>31</td>\n",
       "      <td>[RO:0002434, RO:0002200, None, dc:source]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [DISO], [GENE], [GENE]]</td>\n",
       "      <td>18</td>\n",
       "      <td>[RO:0002434, RO:0002200, RO:0002200, RO:HOM0000011]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [DISO], [DISO], [GENE]]</td>\n",
       "      <td>16</td>\n",
       "      <td>[RO:0002434, RO:0002200, RO:0002200, RO:0002607]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>16</td>\n",
       "      <td>[RO:0002434, RO:HOM0000017, RO:HOM0000020, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>14</td>\n",
       "      <td>[RO:0002434, RO:0002325, RO:0002331, RO:0002331]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [ANAT], [GENE], [GENE]]</td>\n",
       "      <td>12</td>\n",
       "      <td>[RO:0002434, RO:0002206, RO:0002206, RO:HOM0000011]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [ANAT], [GENE]]</td>\n",
       "      <td>10</td>\n",
       "      <td>[RO:0002434, RO:0002435, RO:0002206, RO:0002206]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>10</td>\n",
       "      <td>[RO:0002434, RO:0002435, RO:0002331, RO:0002331]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>8</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002435, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [DISO], [GENE]]</td>\n",
       "      <td>6</td>\n",
       "      <td>[RO:0002434, RO:0002325, RO:0002607, RO:0002607]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>6</td>\n",
       "      <td>[RO:0002434, RO:0002325, BFO:0000050, BFO:0000050]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>6</td>\n",
       "      <td>[RO:0002434, RO:HOM0000011, RO:0002327, RO:0002327]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [DISO], [GENE]]</td>\n",
       "      <td>5</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002326, RO:0002607]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [DISO], [GENE], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:0002200, RO:0002200, RO:HOM0000020]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [ANAT], [GENE], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:0002206, RO:0002206, RO:HOM0000017]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [DISO], [GENE], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:0002200, RO:0002326, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:0002325, RO:0002325, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:HOM0000011, BFO:0000050, BFO:0000050]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:0002325, RO:0002327, RO:0002327]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [PHYS], [GENE], [GENE]]</td>\n",
       "      <td>3</td>\n",
       "      <td>[RO:0002434, RO:0002331, dc:source, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [ANAT], [GENE], [GENE]]</td>\n",
       "      <td>2</td>\n",
       "      <td>[RO:0002434, RO:0002206, RO:0002206, RO:HOM0000020]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [PHYS], [DISO], [GENE]]</td>\n",
       "      <td>2</td>\n",
       "      <td>[RO:0002434, RO:0002331, RO:0002331, RO:0002607]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [DISO], [GENE]]</td>\n",
       "      <td>2</td>\n",
       "      <td>[RO:0002434, RO:0002435, RO:0002607, RO:0002607]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>1</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002331, dc:source]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Metapath length                                     Nodes  Paths  \\\n",
       "0   3                [[GENE], [GENE], [GENE], [GENE]]          36      \n",
       "1   3                [[GENE], [GENE], [GENE], [GENE]]          8       \n",
       "2   3                [[GENE], [GENE], [ANAT], [GENE]]          5       \n",
       "3   4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  6534    \n",
       "4   4                [[GENE], [GENE], [GENE], [ANAT], [GENE]]  1502    \n",
       "5   4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  432     \n",
       "6   4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  425     \n",
       "7   4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  400     \n",
       "8   4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  208     \n",
       "9   4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  190     \n",
       "10  4                [[GENE], [GENE], [DISO], [GENE], [GENE]]  163     \n",
       "11  4                [[GENE], [GENE], [ANAT], [GENE], [GENE]]  145     \n",
       "12  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  136     \n",
       "13  4                [[GENE], [GENE], [DISO], [PHYS], [GENE]]  113     \n",
       "14  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  106     \n",
       "15  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  96      \n",
       "16  4                [[GENE], [GENE], [GENE], [ANAT], [GENE]]  94      \n",
       "17  4                [[GENE], [GENE], [PHYS], [GENE], [GENE]]  87      \n",
       "18  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  72      \n",
       "19  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  64      \n",
       "20  4                [[GENE], [GENE], [GENE], [DISO], [GENE]]  60      \n",
       "21  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  56      \n",
       "22  4                [[GENE], [GENE], [GENE], [ANAT], [GENE]]  54      \n",
       "23  4                [[GENE], [GENE], [PHYS], [GENE], [GENE]]  35      \n",
       "24  4                [[GENE], [GENE], [DISO], [PHYS], [GENE]]  31      \n",
       "25  4                [[GENE], [GENE], [DISO], [GENE], [GENE]]  18      \n",
       "26  4                [[GENE], [GENE], [DISO], [DISO], [GENE]]  16      \n",
       "27  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  16      \n",
       "28  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  14      \n",
       "29  4                [[GENE], [GENE], [ANAT], [GENE], [GENE]]  12      \n",
       "30  4                [[GENE], [GENE], [GENE], [ANAT], [GENE]]  10      \n",
       "31  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  10      \n",
       "32  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  8       \n",
       "33  4                [[GENE], [GENE], [GENE], [DISO], [GENE]]  6       \n",
       "34  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  6       \n",
       "35  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  6       \n",
       "36  4                [[GENE], [GENE], [GENE], [DISO], [GENE]]  5       \n",
       "37  4                [[GENE], [GENE], [DISO], [GENE], [GENE]]  4       \n",
       "38  4                [[GENE], [GENE], [ANAT], [GENE], [GENE]]  4       \n",
       "39  4                [[GENE], [GENE], [DISO], [GENE], [GENE]]  4       \n",
       "40  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  4       \n",
       "41  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  4       \n",
       "42  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  4       \n",
       "43  4                [[GENE], [GENE], [PHYS], [GENE], [GENE]]  3       \n",
       "44  4                [[GENE], [GENE], [ANAT], [GENE], [GENE]]  2       \n",
       "45  4                [[GENE], [GENE], [PHYS], [DISO], [GENE]]  2       \n",
       "46  4                [[GENE], [GENE], [GENE], [DISO], [GENE]]  2       \n",
       "47  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  1       \n",
       "\n",
       "                                                 Relations  \n",
       "0   [RO:0002434, RO:0002434, RO:0002434]                    \n",
       "1   [RO:0002434, RO:HOM0000011, RO:0002434]                 \n",
       "2   [RO:0002434, RO:0002206, RO:0002206]                    \n",
       "3   [RO:0002434, RO:0002434, RO:0002434, RO:0002434]        \n",
       "4   [RO:0002434, RO:0002434, RO:0002206, RO:0002206]        \n",
       "5   [RO:0002434, RO:0002434, BFO:0000050, BFO:0000050]      \n",
       "6   [RO:0002434, RO:0002434, RO:0002331, RO:0002331]        \n",
       "7   [RO:0002434, RO:HOM0000011, RO:0002434, RO:0002434]     \n",
       "8   [RO:0002434, RO:HOM0000011, RO:HOM0000011, RO:0002434]  \n",
       "9   [RO:0002434, RO:0002325, RO:0002434, RO:0002434]        \n",
       "10  [RO:0002434, RO:0002200, RO:0002200, RO:0002434]        \n",
       "11  [RO:0002434, RO:0002206, RO:0002206, RO:0002434]        \n",
       "12  [RO:0002434, RO:0002434, RO:0002325, RO:0002434]        \n",
       "13  [RO:0002434, RO:0002200, None, RO:0002331]              \n",
       "14  [RO:0002434, RO:0002434, RO:0002327, RO:0002327]        \n",
       "15  [RO:0002434, RO:0002434, RO:HOM0000011, RO:0002434]     \n",
       "16  [RO:0002434, RO:HOM0000011, RO:0002206, RO:0002206]     \n",
       "17  [RO:0002434, RO:0002331, RO:0002331, RO:0002434]        \n",
       "18  [RO:0002434, RO:0002435, RO:0002434, RO:0002434]        \n",
       "19  [RO:0002434, RO:HOM0000017, RO:HOM0000017, RO:0002434]  \n",
       "20  [RO:0002434, RO:0002434, RO:0002607, RO:0002607]        \n",
       "21  [RO:0002434, RO:0002434, RO:0002434, RO:HOM0000011]     \n",
       "22  [RO:0002434, RO:0002325, RO:0002206, RO:0002206]        \n",
       "23  [RO:0002434, RO:0002327, RO:0002327, RO:0002434]        \n",
       "24  [RO:0002434, RO:0002200, None, dc:source]               \n",
       "25  [RO:0002434, RO:0002200, RO:0002200, RO:HOM0000011]     \n",
       "26  [RO:0002434, RO:0002200, RO:0002200, RO:0002607]        \n",
       "27  [RO:0002434, RO:HOM0000017, RO:HOM0000020, RO:0002434]  \n",
       "28  [RO:0002434, RO:0002325, RO:0002331, RO:0002331]        \n",
       "29  [RO:0002434, RO:0002206, RO:0002206, RO:HOM0000011]     \n",
       "30  [RO:0002434, RO:0002435, RO:0002206, RO:0002206]        \n",
       "31  [RO:0002434, RO:0002435, RO:0002331, RO:0002331]        \n",
       "32  [RO:0002434, RO:0002434, RO:0002435, RO:0002434]        \n",
       "33  [RO:0002434, RO:0002325, RO:0002607, RO:0002607]        \n",
       "34  [RO:0002434, RO:0002325, BFO:0000050, BFO:0000050]      \n",
       "35  [RO:0002434, RO:HOM0000011, RO:0002327, RO:0002327]     \n",
       "36  [RO:0002434, RO:0002434, RO:0002326, RO:0002607]        \n",
       "37  [RO:0002434, RO:0002200, RO:0002200, RO:HOM0000020]     \n",
       "38  [RO:0002434, RO:0002206, RO:0002206, RO:HOM0000017]     \n",
       "39  [RO:0002434, RO:0002200, RO:0002326, RO:0002434]        \n",
       "40  [RO:0002434, RO:0002325, RO:0002325, RO:0002434]        \n",
       "41  [RO:0002434, RO:HOM0000011, BFO:0000050, BFO:0000050]   \n",
       "42  [RO:0002434, RO:0002325, RO:0002327, RO:0002327]        \n",
       "43  [RO:0002434, RO:0002331, dc:source, RO:0002434]         \n",
       "44  [RO:0002434, RO:0002206, RO:0002206, RO:HOM0000020]     \n",
       "45  [RO:0002434, RO:0002331, RO:0002331, RO:0002607]        \n",
       "46  [RO:0002434, RO:0002435, RO:0002607, RO:0002607]        \n",
       "47  [RO:0002434, RO:0002434, RO:0002331, dc:source]         "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# summary table of metapaths\n",
    "res_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### template refinement (only regulation edge)\n",
    "* **Query 12: ( L = 4 ) ngly1--TF-[regulatory]->gene--phys--aqp1 **\n",
    "    * filter specific regulatory types on specific edges \n",
    "    * filter directionality\n",
    "    * filter node type on third node, otherwise runtime is long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 15628\n",
      "CPU times: user 2.79 ms, sys: 273 µs, total: 3.06 ms\n",
      "Wall time: 31.4 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 12\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)--(tf:GENE)-[i1:`RO:0002434`]->(:GENE)--(:PHYS)--(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 AND toLower(i1.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Query 11: ( L = 3 ) ngly1--TF-[regulatory]->gene--aqp1 **\n",
    "    * filter specific regulatory types on specific edges \n",
    "    * filter directionality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 338\n",
      "CPU times: user 2.36 ms, sys: 112 µs, total: 2.48 ms\n",
      "Wall time: 1.06 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 11\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)--(tf:GENE)-[i1:`RO:0002434`]->(:GENE)--(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 AND toLower(i1.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mitali queries\n",
    "### 14 Jan 2019 meeting \n",
    "\n",
    "Here are the details of the querry we talked about yesterday;\n",
    "\n",
    "\n",
    "Look for gene-gene interactions linking NGLY1 and AQPs through the following genes;\n",
    "\n",
    "* ATF1 (ENSG00000123268)\n",
    "    * https://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000123268;r=12:50763710-50821122\n",
    "\n",
    "* CREB1 (ENSG00000118260)\n",
    "* PRKACA (ENSG00000072062)\n",
    "* cyclic adenosine monophosphate (cAMP)\n",
    "* **cAMP** is not a protein so there is no identification number for it. \n",
    "\n",
    "_Note_: Protein kinase A has multiple subunits. Each subunit has a different identifier. I have provided you with the gene identifier for its catalytic subunit PRKACA.\n",
    "\n",
    "\n",
    "**QUERY TOPOLOGY**:\n",
    "\n",
    "*  NGLY1-interacts-(gene)-interacts-(cAMP|ATF1|CREB1|PKA),  with length n=1/2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GENE IDs\n",
    "\n",
    "I use BioThings to retrieve gene IDs [http://mygene.info/v3/api#/annotation/get_gene__geneid_]\n",
    "\n",
    "* ATF1: \"ATF1\", \"ENSEMBL:ENSG00000123268\", \"HGNC:783\", \"NCBIGene:466\", \"MGI:1298366\", \"FlyBase:FBgn0265784\", \n",
    "    * [http://mygene.info/v3/gene/ENSG00000123268]\n",
    "    * curl -X GET \"http://mygene.info/v3/gene/ENSG00000123268\" -H \"accept: application/json\"\n",
    "    * \"alias\": \"EWS-ATF1\",\"FUS/ATF-1\",\"TREB36\"\n",
    "\n",
    "* CREB1: \"CREB1\", \"ENSEMBL:ENSG00000118260\", \"HGNC:2345\", \"NCBIGene:1385\", \"MGI:88494\", \"FlyBase:FBgn0265784\", \n",
    "    * [http://mygene.info/v3/gene/ENSG00000118260]\n",
    "    * curl -X GET \"http://mygene.info/v3/gene/ENSG00000118260\" -H \"accept: application/json\"\n",
    "    * \"alias\": \"CREB\",\"CREB-1\"\n",
    "  \n",
    "* PKA_PRKACA: \"PRKACA\", \"ENSEMBL:ENSG00000072062\", \"HGNC:9380\", \"NCBIGene:5566\", \"MGI:97592\", \"FlyBase:FBgn0039796\", \"FlyBase:FBgn0000274\", \"FlyBase:FBgn0000273\", \n",
    "    * [http://mygene.info/v3/gene/ENSG00000072062]\n",
    "    * curl -X GET \"http://mygene.info/v3/gene/ENSG00000072062\" -H \"accept: application/json\"\n",
    "    * Fly have orthology type of relation \"O\" (mouse is LDO)\n",
    "    * \"alias\": \"PKACA\",\"PPNAD4\"\n",
    "    \n",
    " _Note_: Fly genes are more distant, broader orthology relation type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 92\n",
      "CPU times: user 0 ns, sys: 1.63 ms, total: 1.63 ms\n",
      "Wall time: 80.8 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)-[:`RO:0002434`]-(:GENE)-[:`RO:0002434`]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:783' \n",
    "        \n",
    "        AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 12042\n",
      "CPU times: user 2.64 ms, sys: 0 ns, total: 2.64 ms\n",
      "Wall time: 2.6 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)-[:`RO:0002434`]-(:GENE)-[:`RO:0002434`]-(:GENE)-[:`RO:0002434`]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:783' \n",
    "        \n",
    "        AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 12134\n",
      "CPU times: user 2.72 ms, sys: 359 µs, total: 3.08 ms\n",
      "Wall time: 484 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)-[:`RO:0002434`]-(:GENE)-[:`RO:0002434`*..2]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:783' \n",
    "        \n",
    "        AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 17005\n",
      "CPU times: user 0 ns, sys: 2.73 ms, total: 2.73 ms\n",
      "Wall time: 677 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)-[:`RO:0002434`]-(:GENE)-[*..2]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:783' \n",
    "        \n",
    "        AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 17069\n",
      "CPU times: user 0 ns, sys: 2.34 ms, total: 2.34 ms\n",
      "Wall time: 696 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)--(:GENE)-[*..2]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:783' \n",
    "        \n",
    "        AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Query: ( L = 3 ) ngly1--TF-[regulatory]->gene--atf1 **\n",
    "    * filter specific regulatory types on specific edges \n",
    "    * filter directionality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 141\n",
      "CPU times: user 1.9 ms, sys: 185 µs, total: 2.09 ms\n",
      "Wall time: 1.83 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)-[i1:`RO:0002434`]-(tf:GENE)-[*..2]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:783' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 AND toLower(i1.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 0\n",
      "CPU times: user 1.78 ms, sys: 176 µs, total: 1.95 ms\n",
      "Wall time: 29.2 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)--(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:783' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 \n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 92\n",
      "CPU times: user 1.65 ms, sys: 162 µs, total: 1.81 ms\n",
      "Wall time: 51.8 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)-[i1:`RO:0002434`]-(:GENE)-[i2:`RO:0002434`]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:783' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 \n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 12042\n",
      "CPU times: user 125 µs, sys: 3.22 ms, total: 3.35 ms\n",
      "Wall time: 2.78 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)-[i1:`RO:0002434`]-(:GENE)-[i2:`RO:0002434`]-(:GENE)-[i3:`RO:0002434`]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:783' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 \n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
