{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cypher Queries for Determining Regulatory Paths\n",
    "*Núria Queralt Rosinach, Andrew Su*\n",
    "\n",
    "**Freeze Lab meeting, January 2019**\n",
    "\n",
    "## Overview\n",
    "NGLY1 - AQP1 **regulatory review** (*NGLY1 v3.2*)\n",
    "\n",
    "## Servers \n",
    "\n",
    "    * Local: bolt://kylo.scripps.edu:7690"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from neo4j.v1 import GraphDatabase, basic_auth\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)  # or 1000\n",
    "pd.set_option('display.max_rows', None)  # or 1000\n",
    "pd.set_option('display.max_colwidth', -1)  # or 199"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def runQuery( driver, query ):\n",
    "    '''\n",
    "    This function runs the query onto the database and returns the result.\n",
    "    in: cypher query string\n",
    "    out: neo4j query result object\n",
    "    '''\n",
    "    \n",
    "    with driver.session() as session:\n",
    "        result = session.run('' + query + '')\n",
    "        \n",
    "    return result\n",
    "\n",
    "\n",
    "def parseNode( node ):\n",
    "    '''\n",
    "    This function parses the information gathered in the node data structure object resulting after querying neo4j.\n",
    "        in: node record neo4j object\n",
    "        out: node as dict\n",
    "    '''\n",
    "    \n",
    "    n = dict()\n",
    "    n[\"idx\"] = int(node.id)\n",
    "    n[\"type\"] = list(node.labels)[0]\n",
    "    n[\"id\"] = str(node.properties['id'])\n",
    "    n[\"preflabel\"] = str(node.properties['preflabel'])\n",
    "    n[\"name\"] = str(node.properties['name'])\n",
    "    n[\"description\"] = str(node.properties['description'])\n",
    "\n",
    "    return n\n",
    "\n",
    "\n",
    "def parsePath( path ):\n",
    "    '''\n",
    "    This function parsers the information gathered in the path data structure object resulting after querying neo4j.\n",
    "        in: path record neo4j object\n",
    "        out: path as dict\n",
    "    '''\n",
    "    \n",
    "    out = {}\n",
    "    out['Nodes'] = []\n",
    "    for node in path['path'].nodes:\n",
    "        n = {}\n",
    "        n['idx'] = int(node.id)\n",
    "        n['type'] = list(node.labels)[0]\n",
    "        n['id'] = str(node.properties['id'])\n",
    "        n['preflabel'] = str(node.properties['preflabel'])\n",
    "        n['name'] = str(node.properties['name'])\n",
    "        n['description'] = str(node.properties['description'])\n",
    "        out['Nodes'].append(n)\n",
    "    out['Edges'] = []\n",
    "    for edge in path['path'].relationships:\n",
    "        e = {}\n",
    "        e['idx'] = int(edge.id)\n",
    "        e['start_node'] = int(edge.start)\n",
    "        e['end_node'] = int(edge.end)\n",
    "        e['type'] = str(edge.type)\n",
    "        e['property_label'] = str(edge.properties['property_label'])\n",
    "        e['property_uri'] = str(edge.properties['property_uri'])\n",
    "        e['reference_uri'] = str(edge.properties['reference_uri'])\n",
    "        e['reference_date'] = str(edge.properties['reference_date'])\n",
    "        e['reference_supporting_text'] = str(edge.properties['reference_supporting_text'])\n",
    "        out['Edges'].append(e)\n",
    "        \n",
    "    return out"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initialize neo4j"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "driver = GraphDatabase.driver(\"bolt://kylo.scripps.edu:7690\", auth=basic_auth(\"neo4j\", \"xena\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Question\n",
    "## Query topology graph\n",
    "## Table of summary\n",
    "## Graph of paths\n",
    "## Explore paths=> executable cypher query"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TRANSCRIPTOME ANALYSIS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Common TFs for human RNA transcriptome\n",
    "* **Query web 1: < NGLY1 -[Human RNA expression (Freeze)]- gene <-- TF > order by freq**\n",
    "    * ngly1-[interacts with]-gene<-tf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12, 5)\n",
      "CPU times: user 0 ns, sys: 3.7 ms, total: 3.7 ms\n",
      "Wall time: 994 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)-[i1:`RO:0002434`]-(fly_gene:GENE)-[:`RO:HOM0000020`]-(gene:GENE)<-[i2:`RO:0002434`]-(tf:GENE)\n",
    "\n",
    "        WHERE source.id = 'FlyBase:FBgn0033050' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,i2,tf,gene,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 \n",
    "        \n",
    "        AND toLower(i1.reference_uri) =~ '.*pubmed/29346549.*' \n",
    "        \n",
    "        AND toLower(i2.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN DISTINCT tf.id as id, tf.preflabel as symbol, tf.name as name, tf.description as description, count(distinct gene.preflabel) as freq\n",
    "        \n",
    "        ORDER BY freq DESC\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "out_l = list()\n",
    "for record in result:\n",
    "    out_l.append({\n",
    "        'TF_id': record['id'], \n",
    "        'TF_symbol': record['symbol'],\n",
    "        'TF_name': record['name'],\n",
    "        'TF_description': record['description'],\n",
    "        'Frequency': record['freq']\n",
    "    })\n",
    "    \n",
    "res_df = pd.DataFrame(out_l)\n",
    "print(res_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Frequency</th>\n",
       "      <th>TF_id</th>\n",
       "      <th>TF_symbol</th>\n",
       "      <th>TF_name</th>\n",
       "      <th>TF_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>HGNC:7996</td>\n",
       "      <td>NRF1</td>\n",
       "      <td>nuclear respiratory factor 1</td>\n",
       "      <td>This gene encodes a protein that homodimerizes and functions as a transcription factor which activates the expression of some key metabolic genes regulating cellular growth and nuclear genes required for respiration, heme biosynthesis, and mitochondrial DNA transcription and replication. The protein has also been associated with the regulation of neurite outgrowth. Alternative splicing results in multiple transcript variants. Confusion has occurred in bibliographic databases due to the shared symbol of NRF1 for this gene and for 'nuclear factor (erythroid-derived 2)-like 1' which has an official symbol of NFE2L1. [provided by RefSeq, May 2014].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>HGNC:6204</td>\n",
       "      <td>JUN</td>\n",
       "      <td>Jun proto-oncogene, AP-1 transcription factor subunit</td>\n",
       "      <td>This gene is the putative transforming gene of avian sarcoma virus 17. It encodes a protein which is highly similar to the viral protein, and which interacts directly with specific target DNA sequences to regulate gene expression. This gene is intronless and is mapped to 1p32-p31, a chromosomal region involved in both translocations and deletions in human malignancies. [provided by RefSeq, Jul 2008].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7782</td>\n",
       "      <td>NFE2L2</td>\n",
       "      <td>nuclear factor, erythroid 2 like 2</td>\n",
       "      <td>This gene encodes a transcription factor which is a member of a small family of basic leucine zipper (bZIP) proteins. The encoded transcription factor regulates genes which contain antioxidant response elements (ARE) in their promoters; many of these genes encode proteins involved in response to injury and inflammation which includes the production of free radicals. Multiple transcript variants encoding different isoforms have been characterized for this gene. [provided by RefSeq, Sep 2015].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:6780</td>\n",
       "      <td>MAFF</td>\n",
       "      <td>MAF bZIP transcription factor F</td>\n",
       "      <td>The protein encoded by this gene is a basic leucine zipper (bZIP) transcription factor that lacks a transactivation domain. It is known to bind the US-2 DNA element in the promoter of the oxytocin receptor (OTR) gene and most likely heterodimerizes with other leucine zipper-containing proteins to enhance expression of the OTR gene during term pregnancy. The encoded protein can also form homodimers, and since it lacks a transactivation domain, the homodimer may act as a repressor of transcription. This gene may also be involved in the cellular stress response. Multiple transcript variants encoding two different isoforms have been found for this gene. [provided by RefSeq, Jun 2009].</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7780</td>\n",
       "      <td>NFE2</td>\n",
       "      <td>nuclear factor, erythroid 2</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Frequency      TF_id TF_symbol  \\\n",
       "0  5          HGNC:7996  NRF1       \n",
       "1  5          HGNC:6204  JUN        \n",
       "2  3          HGNC:7782  NFE2L2     \n",
       "3  3          HGNC:6780  MAFF       \n",
       "4  3          HGNC:7780  NFE2       \n",
       "\n",
       "                                                 TF_name  \\\n",
       "0  nuclear respiratory factor 1                            \n",
       "1  Jun proto-oncogene, AP-1 transcription factor subunit   \n",
       "2  nuclear factor, erythroid 2 like 2                      \n",
       "3  MAF bZIP transcription factor F                         \n",
       "4  nuclear factor, erythroid 2                             \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      TF_description  \n",
       "0  This gene encodes a protein that homodimerizes and functions as a transcription factor which activates the expression of some key metabolic genes regulating cellular growth and nuclear genes required for respiration, heme biosynthesis, and mitochondrial DNA transcription and replication. The protein has also been associated with the regulation of neurite outgrowth. Alternative splicing results in multiple transcript variants. Confusion has occurred in bibliographic databases due to the shared symbol of NRF1 for this gene and for 'nuclear factor (erythroid-derived 2)-like 1' which has an official symbol of NFE2L1. [provided by RefSeq, May 2014].                                       \n",
       "1  This gene is the putative transforming gene of avian sarcoma virus 17. It encodes a protein which is highly similar to the viral protein, and which interacts directly with specific target DNA sequences to regulate gene expression. This gene is intronless and is mapped to 1p32-p31, a chromosomal region involved in both translocations and deletions in human malignancies. [provided by RefSeq, Jul 2008].                                                                                                                                                                                                                                                                                                \n",
       "2  This gene encodes a transcription factor which is a member of a small family of basic leucine zipper (bZIP) proteins. The encoded transcription factor regulates genes which contain antioxidant response elements (ARE) in their promoters; many of these genes encode proteins involved in response to injury and inflammation which includes the production of free radicals. Multiple transcript variants encoding different isoforms have been characterized for this gene. [provided by RefSeq, Sep 2015].                                                                                                                                                                                                   \n",
       "3  The protein encoded by this gene is a basic leucine zipper (bZIP) transcription factor that lacks a transactivation domain. It is known to bind the US-2 DNA element in the promoter of the oxytocin receptor (OTR) gene and most likely heterodimerizes with other leucine zipper-containing proteins to enhance expression of the OTR gene during term pregnancy. The encoded protein can also form homodimers, and since it lacks a transactivation domain, the homodimer may act as a repressor of transcription. This gene may also be involved in the cellular stress response. Multiple transcript variants encoding two different isoforms have been found for this gene. [provided by RefSeq, Jun 2009].  \n",
       "4  NA                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Summary table\n",
    "res_df[['Frequency', 'TF_id', 'TF_symbol', 'TF_name', 'TF_description']].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RNA gene clusters by TF and Pathway (pathways associated with gene sets)\n",
    "    * Cluster RNA genes-common TF clusters by pathway (pathways associated with RNA genes in gene sets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12, 6)\n",
      "CPU times: user 5.66 ms, sys: 133 µs, total: 5.79 ms\n",
      "Wall time: 5.74 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 20\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)-[i1:`RO:0002434`]->(fly_g:GENE)-[:`RO:HOM0000020`]-(g:GENE)<-[i2:`RO:0002434`]-(tf:GENE), (g:GENE)-[i]-(pw:PHYS)\n",
    "\n",
    "        WHERE source.id = 'FlyBase:FBgn0033050' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,i2,g,tf,i,pw,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 \n",
    "        \n",
    "        AND toLower(i1.reference_uri) =~ '.*pubmed/29346549.*'  \n",
    "        \n",
    "        AND toLower(i2.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "        \n",
    "        AND toLower(i.property_label) <> 'enables'\n",
    "\n",
    "        RETURN DISTINCT tf.preflabel as TF_symbol, tf.name as TF_name, \n",
    "                        collect(DISTINCT g.preflabel) as geneset, count(distinct g.preflabel) as genes,\n",
    "                        collect(DISTINCT pw.preflabel) as pathway, count(distinct pw.preflabel) as pathways\n",
    "        \n",
    "        ORDER BY genes DESC\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "out_l = list()\n",
    "for record in result:\n",
    "    out_l.append({'TF symbol': record['TF_symbol'],\n",
    "                  'TF name': record['TF_name'],\n",
    "                  'Target geneset': record['geneset'],\n",
    "                  'Total target genes': record['genes'],\n",
    "                  'Pathway': record['pathway'],\n",
    "                  'Total pathways': record['pathways']\n",
    "                 })\n",
    "    \n",
    "res_df = pd.DataFrame(out_l)\n",
    "print(res_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway</th>\n",
       "      <th>TF name</th>\n",
       "      <th>TF symbol</th>\n",
       "      <th>Target geneset</th>\n",
       "      <th>Total pathways</th>\n",
       "      <th>Total target genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[extracellular exosome, Metabolism, nucleus, Metabolism of proteins, Post-translational protein modification, transmembrane transport, Transmembrane transport of small molecules, positive regulation of transcription from RNA polymerase II promoter, positive regulation of transcription, DNA-templated, regulation of transcription from RNA polymerase II promoter, negative regulation of apoptotic process, protein complex, transcription from RNA polymerase II promoter, response to drug, cellular response to UV, cellular response to hypoxia, positive regulation of fibroblast proliferation]</td>\n",
       "      <td>nuclear respiratory factor 1</td>\n",
       "      <td>NRF1</td>\n",
       "      <td>[GART, PSMA4, HIBADH, LIG3, MYC]</td>\n",
       "      <td>17</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[nucleus, Metabolism of proteins, positive regulation of transcription from RNA polymerase II promoter, positive regulation of transcription, DNA-templated, regulation of transcription from RNA polymerase II promoter, negative regulation of apoptotic process, Post-translational protein modification, protein complex, transcription from RNA polymerase II promoter, response to drug, cellular response to UV, cellular response to hypoxia, positive regulation of fibroblast proliferation, transmembrane transport, Transmembrane transport of small molecules, Metabolism, extracellular exosome, endoplasmic reticulum membrane, neuron projection, dendrite, anatomical structure morphogenesis, plasma membrane, integral component of plasma membrane, cell surface]</td>\n",
       "      <td>Jun proto-oncogene, AP-1 transcription factor subunit</td>\n",
       "      <td>JUN</td>\n",
       "      <td>[MYC, PSMC3, HPD, TH, SLC6A2]</td>\n",
       "      <td>24</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[nucleus, Metabolism of proteins, positive regulation of transcription from RNA polymerase II promoter, positive regulation of transcription, DNA-templated, regulation of transcription from RNA polymerase II promoter, negative regulation of apoptotic process, Post-translational protein modification, protein complex, transcription from RNA polymerase II promoter, response to drug, cellular response to UV, cellular response to hypoxia, positive regulation of fibroblast proliferation, extracellular exosome, Metabolism, endoplasmic reticulum membrane]</td>\n",
       "      <td>MAF bZIP transcription factor F</td>\n",
       "      <td>MAFF</td>\n",
       "      <td>[MYC, PSMG2, HPD]</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[nucleus, endoplasmic reticulum, extracellular exosome, positive regulation of transcription from RNA polymerase II promoter, negative regulation of apoptotic process, Metabolism of proteins, positive regulation of transcription, DNA-templated, regulation of transcription from RNA polymerase II promoter, Post-translational protein modification, protein complex, transcription from RNA polymerase II promoter, response to drug, cellular response to UV, cellular response to hypoxia, positive regulation of fibroblast proliferation]</td>\n",
       "      <td>nuclear factor, erythroid 2 like 2</td>\n",
       "      <td>NFE2L2</td>\n",
       "      <td>[POMP, SQSTM1, MYC]</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[nucleus, Metabolism of proteins, extracellular exosome, Post-translational protein modification, transmembrane transport, Transmembrane transport of small molecules, Metabolism, integral component of membrane, endoplasmic reticulum, endoplasmic reticulum membrane, brush border, positive regulation of transcription from RNA polymerase II promoter, negative regulation of apoptotic process]</td>\n",
       "      <td>nuclear factor, erythroid 2</td>\n",
       "      <td>NFE2</td>\n",
       "      <td>[PSMA1, LCTL, SQSTM1]</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 Pathway  \\\n",
       "0  [extracellular exosome, Metabolism, nucleus, Metabolism of proteins, Post-translational protein modification, transmembrane transport, Transmembrane transport of small molecules, positive regulation of transcription from RNA polymerase II promoter, positive regulation of transcription, DNA-templated, regulation of transcription from RNA polymerase II promoter, negative regulation of apoptotic process, protein complex, transcription from RNA polymerase II promoter, response to drug, cellular response to UV, cellular response to hypoxia, positive regulation of fibroblast proliferation]                                                                                                                                                                          \n",
       "1  [nucleus, Metabolism of proteins, positive regulation of transcription from RNA polymerase II promoter, positive regulation of transcription, DNA-templated, regulation of transcription from RNA polymerase II promoter, negative regulation of apoptotic process, Post-translational protein modification, protein complex, transcription from RNA polymerase II promoter, response to drug, cellular response to UV, cellular response to hypoxia, positive regulation of fibroblast proliferation, transmembrane transport, Transmembrane transport of small molecules, Metabolism, extracellular exosome, endoplasmic reticulum membrane, neuron projection, dendrite, anatomical structure morphogenesis, plasma membrane, integral component of plasma membrane, cell surface]   \n",
       "2  [nucleus, Metabolism of proteins, positive regulation of transcription from RNA polymerase II promoter, positive regulation of transcription, DNA-templated, regulation of transcription from RNA polymerase II promoter, negative regulation of apoptotic process, Post-translational protein modification, protein complex, transcription from RNA polymerase II promoter, response to drug, cellular response to UV, cellular response to hypoxia, positive regulation of fibroblast proliferation, extracellular exosome, Metabolism, endoplasmic reticulum membrane]                                                                                                                                                                                                               \n",
       "3  [nucleus, endoplasmic reticulum, extracellular exosome, positive regulation of transcription from RNA polymerase II promoter, negative regulation of apoptotic process, Metabolism of proteins, positive regulation of transcription, DNA-templated, regulation of transcription from RNA polymerase II promoter, Post-translational protein modification, protein complex, transcription from RNA polymerase II promoter, response to drug, cellular response to UV, cellular response to hypoxia, positive regulation of fibroblast proliferation]                                                                                                                                                                                                                                    \n",
       "4  [nucleus, Metabolism of proteins, extracellular exosome, Post-translational protein modification, transmembrane transport, Transmembrane transport of small molecules, Metabolism, integral component of membrane, endoplasmic reticulum, endoplasmic reticulum membrane, brush border, positive regulation of transcription from RNA polymerase II promoter, negative regulation of apoptotic process]                                                                                                                                                                                                                                                                                                                                                                                 \n",
       "\n",
       "                                                 TF name TF symbol  \\\n",
       "0  nuclear respiratory factor 1                           NRF1       \n",
       "1  Jun proto-oncogene, AP-1 transcription factor subunit  JUN        \n",
       "2  MAF bZIP transcription factor F                        MAFF       \n",
       "3  nuclear factor, erythroid 2 like 2                     NFE2L2     \n",
       "4  nuclear factor, erythroid 2                            NFE2       \n",
       "\n",
       "                     Target geneset  Total pathways  Total target genes  \n",
       "0  [GART, PSMA4, HIBADH, LIG3, MYC]  17              5                   \n",
       "1  [MYC, PSMC3, HPD, TH, SLC6A2]     24              5                   \n",
       "2  [MYC, PSMG2, HPD]                 16              3                   \n",
       "3  [POMP, SQSTM1, MYC]               15              3                   \n",
       "4  [PSMA1, LCTL, SQSTM1]             13              3                   "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Summary table\n",
    "res_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NETWORK HYPOTHESES"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### template that involves genes and pathways\n",
    "* **Query 1: ( L = 3 ) (ngly1-ppi-pw-aqp1) ngly1-[interacts with]-gene--phys--aqp1 **\n",
    "    * without degree filters\n",
    "    * this query in graph v2 gives 0 paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total paths: 111\n",
      "CPU times: user 1.49 ms, sys: 0 ns, total: 1.49 ms\n",
      "Wall time: 95.5 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 1\n",
    "with driver.session() as session:\n",
    "    result = session.run(\n",
    "    \"\"\"\n",
    "    MATCH path=(source:GENE)-[:`RO:0002434`]-(:GENE)--(:PHYS)--(target:GENE)\n",
    "\n",
    "    WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "    WITH path,\n",
    "\n",
    "    [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked,\n",
    "\n",
    "    [r IN relationships(path) WHERE r.property_label IN ['in paralogy relationship with','in orthology relationship with','colocalizes with']] AS edges_marked\n",
    "\n",
    "    WHERE size(nodes_marked) = 0 AND size(edges_marked) = 0\n",
    "\n",
    "    RETURN count(distinct path) as paths\n",
    "    \"\"\"\n",
    "    )\n",
    "\n",
    "# print result\n",
    "for record in result:\n",
    "    print('Total paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### template rna-reg L=3 min\n",
    "* **Query 14: TF-pw (L=4) NGLY1--RNA_GENE--TF--PHYS--AQP1**\n",
    "    * filter expression edge \n",
    "    * filter TF regulatory edge and directionality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4.57 ms, sys: 0 ns, total: 4.57 ms\n",
      "Wall time: 3.45 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 14\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE {id: 'FlyBase:FBgn0033050'})-[i1:`RO:0002434`]-(fly_g:GENE)-[:`RO:HOM0000020`]-(rna:GENE)<-[i2:`RO:0002434`]-(tf:GENE)--(pw:PHYS)--(target:GENE {id: 'HGNC:633'})\n",
    "\n",
    "        WHERE ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,i2,rna,tf,pw,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 \n",
    "        \n",
    "        AND toLower(i1.reference_uri) =~ '.*pubmed/29346549.*' \n",
    "        \n",
    "        AND toLower(i2.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN count(distinct rna) as rnas, count(distinct tf) as tfs, count(distinct pw) as pws, count(distinct path) as paths\n",
    "        \n",
    "        ORDER BY paths DESC\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "out_l = list()\n",
    "for record in result:\n",
    "    out_l.append({'Expressed_genes': record['rnas'], \n",
    "                  'TFs': record['tfs'],\n",
    "                  'Pathways': record['pws'],\n",
    "                  'Paths': record['paths']})\n",
    "    \n",
    "res_df = pd.DataFrame(out_l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Expressed_genes</th>\n",
       "      <th>Paths</th>\n",
       "      <th>Pathways</th>\n",
       "      <th>TFs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12</td>\n",
       "      <td>74</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Expressed_genes  Paths  Pathways  TFs\n",
       "0  12               74     9         4  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Summary table\n",
    "res_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Query 14.1: TF-pw (L=3) NGLY1--RNA_GENE--TF--AQP1**\n",
    "    * filter expression edge \n",
    "    * filter TF regulatory edge and directionality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.47 ms, sys: 0 ns, total: 2.47 ms\n",
      "Wall time: 355 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 14.1\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE {id: 'FlyBase:FBgn0033050'})-[i1:`RO:0002434`]-(fly_g:GENE)-[:`RO:HOM0000020`]-(rna:GENE)<-[i2:`RO:0002434`]-(tf:GENE)--(target:GENE {id: 'HGNC:633'})\n",
    "\n",
    "        WHERE ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,i2,rna,tf,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 \n",
    "        \n",
    "        AND toLower(i1.reference_uri) =~ '.*pubmed/29346549.*' \n",
    "        \n",
    "        AND toLower(i2.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN count(distinct rna) as rnas, count(distinct tf) as tfs, count(distinct path) as paths\n",
    "        \n",
    "        ORDER BY paths DESC\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "out_l = list()\n",
    "for record in result:\n",
    "    out_l.append({'Expressed_genes': record['rnas'], \n",
    "                  'TFs': record['tfs'],\n",
    "                  'Paths': record['paths']})\n",
    "    \n",
    "res_df = pd.DataFrame(out_l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Expressed_genes</th>\n",
       "      <th>Paths</th>\n",
       "      <th>TFs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Expressed_genes  Paths  TFs\n",
       "0  2                8      2  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# summary table\n",
    "res_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look for a template of interest (only regulation edge): Open query + metapaths \n",
    "* **Query 13: Variable length 2-4 and pattern**\n",
    "    * pattern: ngly1<-regulatory-tf-[<=3]-aqp1 \n",
    "    * ngly1-[interacts with]-gene-[1..3]-aqp1 \n",
    "    * filter specific regulatory types on specific edges "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 4973\n",
      "CPU times: user 2.84 ms, sys: 242 µs, total: 3.08 ms\n",
      "Wall time: 21.7 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 13\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)<-[i1:`RO:0002434`]-(tf:GENE)-[*..3]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 AND toLower(i1.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 6.93 ms, sys: 148 µs, total: 7.08 ms\n",
      "Wall time: 21.7 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 13\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)<-[i1:`RO:0002434`]-(tf:GENE)-[*..3]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 AND toLower(i1.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN DISTINCT extract (x in rels(path) | type(x)) as types, extract (n in nodes(path) | labels(n)) as labels, length(path) as mp_length, count(distinct path) as paths \n",
    "        \n",
    "        ORDER BY mp_length, paths DESC\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "out_l = list()\n",
    "for record in result:\n",
    "    out_l.append({'Nodes': record['labels'], \n",
    "                  'Relations': record['types'],\n",
    "                  'Metapath length': record['mp_length'],\n",
    "                  'Paths': record['paths']})\n",
    "    \n",
    "res_df = pd.DataFrame(out_l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metapath length</th>\n",
       "      <th>Nodes</th>\n",
       "      <th>Paths</th>\n",
       "      <th>Relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>24</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>8</td>\n",
       "      <td>[RO:0002434, RO:HOM0000011, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>[[GENE], [GENE], [ANAT], [GENE]]</td>\n",
       "      <td>5</td>\n",
       "      <td>[RO:0002434, RO:0002206, RO:0002206]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>2493</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002434, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [ANAT], [GENE]]</td>\n",
       "      <td>544</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002206, RO:0002206]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>196</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002331, RO:0002331]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>196</td>\n",
       "      <td>[RO:0002434, RO:HOM0000011, RO:0002434, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [DISO], [PHYS], [GENE]]</td>\n",
       "      <td>182</td>\n",
       "      <td>[RO:0002434, RO:0002200, None, RO:0002331]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [DISO], [GENE], [GENE]]</td>\n",
       "      <td>171</td>\n",
       "      <td>[RO:0002434, RO:0002200, RO:0002200, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>162</td>\n",
       "      <td>[RO:0002434, RO:0002434, BFO:0000050, BFO:0000050]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>112</td>\n",
       "      <td>[RO:0002434, RO:HOM0000011, RO:HOM0000011, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [ANAT], [GENE], [GENE]]</td>\n",
       "      <td>108</td>\n",
       "      <td>[RO:0002434, RO:0002206, RO:0002206, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>98</td>\n",
       "      <td>[RO:0002434, RO:0002325, RO:0002434, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [PHYS], [GENE], [GENE]]</td>\n",
       "      <td>87</td>\n",
       "      <td>[RO:0002434, RO:0002331, RO:0002331, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>70</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002325, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>64</td>\n",
       "      <td>[RO:0002434, RO:HOM0000017, RO:HOM0000017, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>62</td>\n",
       "      <td>[RO:0002434, RO:0002435, RO:0002434, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>46</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002327, RO:0002327]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [ANAT], [GENE]]</td>\n",
       "      <td>40</td>\n",
       "      <td>[RO:0002434, RO:0002325, RO:0002206, RO:0002206]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [PHYS], [GENE], [GENE]]</td>\n",
       "      <td>35</td>\n",
       "      <td>[RO:0002434, RO:0002327, RO:0002327, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [DISO], [PHYS], [GENE]]</td>\n",
       "      <td>33</td>\n",
       "      <td>[RO:0002434, RO:0002200, None, dc:source]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [ANAT], [GENE]]</td>\n",
       "      <td>32</td>\n",
       "      <td>[RO:0002434, RO:HOM0000011, RO:0002206, RO:0002206]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [DISO], [GENE]]</td>\n",
       "      <td>30</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002607, RO:0002607]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>28</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:HOM0000011, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [DISO], [GENE], [GENE]]</td>\n",
       "      <td>22</td>\n",
       "      <td>[RO:0002434, RO:0002200, RO:0002200, RO:HOM0000011]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [DISO], [DISO], [GENE]]</td>\n",
       "      <td>16</td>\n",
       "      <td>[RO:0002434, RO:0002200, RO:0002200, RO:0002607]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>16</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002434, RO:HOM0000011]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>16</td>\n",
       "      <td>[RO:0002434, RO:HOM0000017, RO:HOM0000020, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [ANAT], [GENE], [GENE]]</td>\n",
       "      <td>10</td>\n",
       "      <td>[RO:0002434, RO:0002206, RO:0002206, RO:HOM0000011]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>10</td>\n",
       "      <td>[RO:0002434, RO:0002435, RO:0002331, RO:0002331]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>10</td>\n",
       "      <td>[RO:0002434, RO:0002325, RO:0002331, RO:0002331]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [DISO], [GENE], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:0002200, RO:0002200, RO:HOM0000020]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [DISO], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:0002325, RO:0002607, RO:0002607]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [ANAT], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:0002435, RO:0002206, RO:0002206]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [DISO], [GENE], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:0002200, RO:0002326, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [GENE], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:0002325, RO:0002325, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:0002325, BFO:0000050, BFO:0000050]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:HOM0000011, BFO:0000050, BFO:0000050]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:0002325, RO:0002327, RO:0002327]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [PHYS], [GENE]]</td>\n",
       "      <td>4</td>\n",
       "      <td>[RO:0002434, RO:HOM0000011, RO:0002327, RO:0002327]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [PHYS], [GENE], [GENE]]</td>\n",
       "      <td>3</td>\n",
       "      <td>[RO:0002434, RO:0002331, dc:source, RO:0002434]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [ANAT], [GENE], [GENE]]</td>\n",
       "      <td>2</td>\n",
       "      <td>[RO:0002434, RO:0002206, RO:0002206, RO:HOM0000020]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [DISO], [GENE]]</td>\n",
       "      <td>2</td>\n",
       "      <td>[RO:0002434, RO:0002434, RO:0002326, RO:0002607]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [PHYS], [DISO], [GENE]]</td>\n",
       "      <td>2</td>\n",
       "      <td>[RO:0002434, RO:0002331, RO:0002331, RO:0002607]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>4</td>\n",
       "      <td>[[GENE], [GENE], [GENE], [DISO], [GENE]]</td>\n",
       "      <td>2</td>\n",
       "      <td>[RO:0002434, RO:0002435, RO:0002607, RO:0002607]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Metapath length                                     Nodes  Paths  \\\n",
       "0   3                [[GENE], [GENE], [GENE], [GENE]]          24      \n",
       "1   3                [[GENE], [GENE], [GENE], [GENE]]          8       \n",
       "2   3                [[GENE], [GENE], [ANAT], [GENE]]          5       \n",
       "3   4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  2493    \n",
       "4   4                [[GENE], [GENE], [GENE], [ANAT], [GENE]]  544     \n",
       "5   4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  196     \n",
       "6   4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  196     \n",
       "7   4                [[GENE], [GENE], [DISO], [PHYS], [GENE]]  182     \n",
       "8   4                [[GENE], [GENE], [DISO], [GENE], [GENE]]  171     \n",
       "9   4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  162     \n",
       "10  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  112     \n",
       "11  4                [[GENE], [GENE], [ANAT], [GENE], [GENE]]  108     \n",
       "12  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  98      \n",
       "13  4                [[GENE], [GENE], [PHYS], [GENE], [GENE]]  87      \n",
       "14  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  70      \n",
       "15  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  64      \n",
       "16  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  62      \n",
       "17  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  46      \n",
       "18  4                [[GENE], [GENE], [GENE], [ANAT], [GENE]]  40      \n",
       "19  4                [[GENE], [GENE], [PHYS], [GENE], [GENE]]  35      \n",
       "20  4                [[GENE], [GENE], [DISO], [PHYS], [GENE]]  33      \n",
       "21  4                [[GENE], [GENE], [GENE], [ANAT], [GENE]]  32      \n",
       "22  4                [[GENE], [GENE], [GENE], [DISO], [GENE]]  30      \n",
       "23  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  28      \n",
       "24  4                [[GENE], [GENE], [DISO], [GENE], [GENE]]  22      \n",
       "25  4                [[GENE], [GENE], [DISO], [DISO], [GENE]]  16      \n",
       "26  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  16      \n",
       "27  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  16      \n",
       "28  4                [[GENE], [GENE], [ANAT], [GENE], [GENE]]  10      \n",
       "29  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  10      \n",
       "30  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  10      \n",
       "31  4                [[GENE], [GENE], [DISO], [GENE], [GENE]]  4       \n",
       "32  4                [[GENE], [GENE], [GENE], [DISO], [GENE]]  4       \n",
       "33  4                [[GENE], [GENE], [GENE], [ANAT], [GENE]]  4       \n",
       "34  4                [[GENE], [GENE], [DISO], [GENE], [GENE]]  4       \n",
       "35  4                [[GENE], [GENE], [GENE], [GENE], [GENE]]  4       \n",
       "36  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  4       \n",
       "37  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  4       \n",
       "38  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  4       \n",
       "39  4                [[GENE], [GENE], [GENE], [PHYS], [GENE]]  4       \n",
       "40  4                [[GENE], [GENE], [PHYS], [GENE], [GENE]]  3       \n",
       "41  4                [[GENE], [GENE], [ANAT], [GENE], [GENE]]  2       \n",
       "42  4                [[GENE], [GENE], [GENE], [DISO], [GENE]]  2       \n",
       "43  4                [[GENE], [GENE], [PHYS], [DISO], [GENE]]  2       \n",
       "44  4                [[GENE], [GENE], [GENE], [DISO], [GENE]]  2       \n",
       "\n",
       "                                                 Relations  \n",
       "0   [RO:0002434, RO:0002434, RO:0002434]                    \n",
       "1   [RO:0002434, RO:HOM0000011, RO:0002434]                 \n",
       "2   [RO:0002434, RO:0002206, RO:0002206]                    \n",
       "3   [RO:0002434, RO:0002434, RO:0002434, RO:0002434]        \n",
       "4   [RO:0002434, RO:0002434, RO:0002206, RO:0002206]        \n",
       "5   [RO:0002434, RO:0002434, RO:0002331, RO:0002331]        \n",
       "6   [RO:0002434, RO:HOM0000011, RO:0002434, RO:0002434]     \n",
       "7   [RO:0002434, RO:0002200, None, RO:0002331]              \n",
       "8   [RO:0002434, RO:0002200, RO:0002200, RO:0002434]        \n",
       "9   [RO:0002434, RO:0002434, BFO:0000050, BFO:0000050]      \n",
       "10  [RO:0002434, RO:HOM0000011, RO:HOM0000011, RO:0002434]  \n",
       "11  [RO:0002434, RO:0002206, RO:0002206, RO:0002434]        \n",
       "12  [RO:0002434, RO:0002325, RO:0002434, RO:0002434]        \n",
       "13  [RO:0002434, RO:0002331, RO:0002331, RO:0002434]        \n",
       "14  [RO:0002434, RO:0002434, RO:0002325, RO:0002434]        \n",
       "15  [RO:0002434, RO:HOM0000017, RO:HOM0000017, RO:0002434]  \n",
       "16  [RO:0002434, RO:0002435, RO:0002434, RO:0002434]        \n",
       "17  [RO:0002434, RO:0002434, RO:0002327, RO:0002327]        \n",
       "18  [RO:0002434, RO:0002325, RO:0002206, RO:0002206]        \n",
       "19  [RO:0002434, RO:0002327, RO:0002327, RO:0002434]        \n",
       "20  [RO:0002434, RO:0002200, None, dc:source]               \n",
       "21  [RO:0002434, RO:HOM0000011, RO:0002206, RO:0002206]     \n",
       "22  [RO:0002434, RO:0002434, RO:0002607, RO:0002607]        \n",
       "23  [RO:0002434, RO:0002434, RO:HOM0000011, RO:0002434]     \n",
       "24  [RO:0002434, RO:0002200, RO:0002200, RO:HOM0000011]     \n",
       "25  [RO:0002434, RO:0002200, RO:0002200, RO:0002607]        \n",
       "26  [RO:0002434, RO:0002434, RO:0002434, RO:HOM0000011]     \n",
       "27  [RO:0002434, RO:HOM0000017, RO:HOM0000020, RO:0002434]  \n",
       "28  [RO:0002434, RO:0002206, RO:0002206, RO:HOM0000011]     \n",
       "29  [RO:0002434, RO:0002435, RO:0002331, RO:0002331]        \n",
       "30  [RO:0002434, RO:0002325, RO:0002331, RO:0002331]        \n",
       "31  [RO:0002434, RO:0002200, RO:0002200, RO:HOM0000020]     \n",
       "32  [RO:0002434, RO:0002325, RO:0002607, RO:0002607]        \n",
       "33  [RO:0002434, RO:0002435, RO:0002206, RO:0002206]        \n",
       "34  [RO:0002434, RO:0002200, RO:0002326, RO:0002434]        \n",
       "35  [RO:0002434, RO:0002325, RO:0002325, RO:0002434]        \n",
       "36  [RO:0002434, RO:0002325, BFO:0000050, BFO:0000050]      \n",
       "37  [RO:0002434, RO:HOM0000011, BFO:0000050, BFO:0000050]   \n",
       "38  [RO:0002434, RO:0002325, RO:0002327, RO:0002327]        \n",
       "39  [RO:0002434, RO:HOM0000011, RO:0002327, RO:0002327]     \n",
       "40  [RO:0002434, RO:0002331, dc:source, RO:0002434]         \n",
       "41  [RO:0002434, RO:0002206, RO:0002206, RO:HOM0000020]     \n",
       "42  [RO:0002434, RO:0002434, RO:0002326, RO:0002607]        \n",
       "43  [RO:0002434, RO:0002331, RO:0002331, RO:0002607]        \n",
       "44  [RO:0002434, RO:0002435, RO:0002607, RO:0002607]        "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# summary table of metapaths\n",
    "res_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### template refinement (only regulation edge)\n",
    "* **Query 12: ( L = 4 ) ngly1--TF-[regulatory]->gene--phys--aqp1 **\n",
    "    * filter specific regulatory types on specific edges \n",
    "    * filter directionality\n",
    "    * filter node type on third node, otherwise runtime is long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 16\n",
      "CPU times: user 2.94 ms, sys: 64 µs, total: 3.01 ms\n",
      "Wall time: 2.54 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 12\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)--(tf:GENE)-[i1:`RO:0002434`]->(:GENE)--(:PHYS)--(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 AND toLower(i1.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Query 11: ( L = 3 ) ngly1--TF-[regulatory]->gene--aqp1 **\n",
    "    * filter specific regulatory types on specific edges \n",
    "    * filter directionality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 6\n",
      "CPU times: user 0 ns, sys: 2.51 ms, total: 2.51 ms\n",
      "Wall time: 267 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Query 11\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)--(tf:GENE)-[i1:`RO:0002434`]->(:GENE)--(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:633' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 AND toLower(i1.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mitali queries\n",
    "### 14 Jan 2019 meeting \n",
    "\n",
    "Here are the details of the querry we talked about yesterday;\n",
    "\n",
    "\n",
    "Look for gene-gene interactions linking NGLY1 and AQPs through the following genes;\n",
    "\n",
    "* ATF1 (ENSG00000123268)\n",
    "    * https://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000123268;r=12:50763710-50821122\n",
    "\n",
    "* CREB1 (ENSG00000118260)\n",
    "* PRKACA (ENSG00000072062)\n",
    "* cyclic adenosine monophosphate (cAMP)\n",
    "* **cAMP** is not a protein so there is no identification number for it. \n",
    "\n",
    "_Note_: Protein kinase A has multiple subunits. Each subunit has a different identifier. I have provided you with the gene identifier for its catalytic subunit PRKACA.\n",
    "\n",
    "\n",
    "**QUERY TOPOLOGY**:\n",
    "\n",
    "*  NGLY1-interacts-(gene)-interacts-(cAMP|ATF1|CREB1|PKA),  with length n=1/2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Query: ( L = 3 ) ngly1--TF-[regulatory]->gene--atf1 **\n",
    "    * filter specific regulatory types on specific edges \n",
    "    * filter directionality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 52\n",
      "CPU times: user 1.77 ms, sys: 141 µs, total: 1.91 ms\n",
      "Wall time: 153 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)-[i1:`RO:0002434`]-(tf:GENE)-[*..2]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:783' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 AND toLower(i1.reference_supporting_text) =~ '.*tftargets.*|.*msigdb.*'\n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 0\n",
      "CPU times: user 1.63 ms, sys: 135 µs, total: 1.77 ms\n",
      "Wall time: 12.9 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)--(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:783' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 \n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 0\n",
      "CPU times: user 1.92 ms, sys: 3 µs, total: 1.92 ms\n",
      "Wall time: 24.7 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)-[i1:`RO:0002434`]-(:GENE)-[i2:`RO:0002434`]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:783' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 \n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paths: 508\n",
      "CPU times: user 3.26 ms, sys: 0 ns, total: 3.26 ms\n",
      "Wall time: 502 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query = (\n",
    "        \"\"\"\n",
    "        MATCH path=(source:GENE)-[i1:`RO:0002434`]-(:GENE)-[i2:`RO:0002434`]-(:GENE)-[i3:`RO:0002434`]-(target:GENE)\n",
    "\n",
    "        WHERE source.id = 'HGNC:17646' AND target.id = 'HGNC:783' AND ALL(x IN nodes(path) WHERE single(y IN nodes(path) WHERE y = x))\n",
    "\n",
    "        WITH path,i1,\n",
    "\n",
    "        [n IN nodes(path) WHERE n.preflabel IN ['cytoplasm','cytosol','nucleus','metabolism','membrane','protein binding','visible','viable','phenotype']] AS nodes_marked\n",
    "\n",
    "        WHERE size(nodes_marked) = 0 \n",
    "\n",
    "        RETURN count(distinct path) as paths\n",
    "        \"\"\"\n",
    ")\n",
    "\n",
    "# run query\n",
    "result = runQuery( driver, query )\n",
    "\n",
    "# parse results\n",
    "for record in result:\n",
    "    print('Paths: {}'.format(record['paths']))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
